---
title: "Introduction to LTRpred"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Introduction}
  %\VignetteEngine{knitr::rmarkdown}
  %\usepackage[utf8]{inputenc}
---

# Table of Contents
1. [Introduction](#introduction)
2. [Getting Started](#getting-started)
3. [Quick Start](#quick-start)
4. [Detailed LTRpred run](#detailed-ltrpred-run)


# Introduction

The `LTRpred` package implements a pipeline and provides an integrated software framework to screen for potentially active LTR transposons _de novo_ in any genomic sequence of interest. For this purpose, this package provides a rich set of analytics tools to allow researchers to quickly 
start annotating and explore their own genomes.

The _de novo_ prediction of LTR transposons in `LTRpred` is based on the command line tools 
[LTRharvest](http://www.zbh.uni-hamburg.de/?id=206) and [LTRdigest](http://www.zbh.uni-hamburg.de/forschung/arbeitsgruppe-genominformatik/software/ltrdigest.html) and extends these search strategies by additional analytics
modules to filter the search space of putative LTR transposons for biologically meaningful candidates.

Please make sure that all command line tools are installed properly before running any `LTRpred` function. Users can find a detailed documentation on how to install all
pre-requisite tools in the [Installation Vignette](https://github.com/HajkD/LTRpred/blob/master/vignettes/Installation.Rmd).

# Getting Started

The rational for implementing `LTRpred` was to implement an R based pipeline combining the most
sensitive, accurate, and conservative state-of-the art LTR retrotransposon
annotation tools and extend their inference by additional analyses and quality filtering steps to
screen for high confidence potentially active elements. Hence, _LTRpred_
aims to provide a high-level _de novo_ prediction infrastructure to generate high quality annotations
of potentially intact LTR retrotransposons. The focus of _LTRpred_ lies on detecting potentially active or recently active
LTR retrotransposons. Therefore, instead of aiming to annotate as much loci in the genome as possible, most functions are optimized to screen for high quality and structurally intact elements that can be studied in transposon mobilization experiments. However, I implemented
all functions generic so that parameters can be changed to detect any form of LTR retrotransposon in any genome.

Internally, _LTRpred_ is based on the `de novo` annotation tools `LTRharvest` and `LTRdigest` which use prior knowledge
about DNA sequence features (also referred to as `Structure-based methods`) such as the homology of Long Terminal Repeats (LTRs), Primer Binding Sites (PBS), _gag_ and _pol_ protein domains, and
Target Site Duplications (TSDs) that are known to enable the process of transposition ([Bergman and Quesneville, 2007](http://bib.oxfordjournals.org/content/8/6/382.long)) to infer LTR retrotransposons in any genome.

Hence, these `de novo` annotation tools are designed to screen the genome systematically and efficiently for
these structural DNA features. __Figure 1__ shows the structural features of LTR retrotransposons
that are used for predicting putative LTR transposons `de novo` in any genome of interest.

![Sequence features of LTR retrotransposons. The structural element (LTR length, LTR similarity, TSD, max. size of full LTR rertrotransposon, PBS length, tRNA binding, protein domain search, etc.) can be mofified and controlled separately by corresponding arguments implemented in the `LTRpred()` function. In addition to controlling the structural features of candidate LTR retrotransposons, a open reading frame prediction, Dfam annotation, copy number clustering, and LTR copy number estimation, and methylation context quantification is performed subsequently after reporting putative LTR retrotransposons.](LTRfeatures.png)

Based on the optimized output of these tools, the `LTRpred` package aims to provide researchers with a maximal flexibility of
parameters that can be used to adjust this structural template to detect any type of LTR retrotransposons. `LTRpred` package allows users to modify a vast range of
parameters to screen for potential LTR transposons, so having this template shown in
__Figure 1__ in mind will help researchers to modify structural parameters in a biologically meaningful manner.

## Quick Start

The fastest way to generate a LTR retrotransposon prediction for a genome of interest (after [installing](Install.Rmd) all prerequisite command line tools) is to use the
`LTRpred()` function and relying on the default parameters. In the following example,
a LTR transposon prediction is performed for parts of the Human Y chromosome.

```r
# load LTRpred package
library(LTRpred)
# de novo LTR transposon prediction for the Human Y chromosome
LTRpred(
    genome.file = system.file("Hsapiens_ChrY.fa", package = "LTRpred"),
    trnas       = system.file("hg38-tRNAs.fa", package = "LTRpred"),
    hmms        = paste0(system.file("HMMs/", package = "LTRpred"), "hmm_*"),
    cluster     = TRUE
)
```

```
Folder 'Hsapiens_ChrY_ltrpred' exists already and will be used...
Starting LTRpred analysis...
Step 1:
Run LTRharvest...
LTRharvest: Generating indexfile Hsapiens_ChrY_ltrharvest/Hsapiens_ChrY_index.fsa with gt suffixerator...
Running LTRharvest and writing results to Hsapiens_ChrY_ltrharvest...
LTRharvest analysis finished!
Step 2:
Generating index file Hsapiens_ChrY_ltrdigest/Hsapiens_ChrY_index_ltrdigest.fsa with suffixerator...
LTRdigest: Sort index file...
Running LTRdigest and writing results to Hsapiens_ChrY_ltrdigest...
LTRdigest analysis finished!
Step 3:
Import LTRdigest Predictions...

Input:  Hsapiens_ChrY_ltrdigest/Hsapiens_ChrY_LTRdigestPrediction.gff  -> Row Number:  168
Remove 'NA' -> New Row Number:  168
(1/8) Filtering for repeat regions has been finished.
(2/8) Filtering for LTR retrotransposons has been finished.
(3/8) Filtering for inverted repeats has been finished.
(4/8) Filtering for LTRs has been finished.
(5/8) Filtering for target site duplication has been finished.
(6/8) Filtering for primer binding site has been finished.
(7/8) Filtering for protein match has been finished.
(8/8) Filtering for RR tract has been finished.
Step 4:
Perform ORF Prediction...
00:00 2.0Mb   100.0% Working

WARNING: Input has lower-case masked sequences

Join ORF Prediction table: nrow(df) = 24 candidates.
unique(ID) = 24 candidates.
unique(orf.id) = 24 candidates.
Perform Clustering of LTR transposons....
Reading file Hsapiens_ChrY_ltrdigest/Hsapiens_ChrY-ltrdigest_complete.fas 100%
340259 nt in 24 seqs, min 5016, max 24131, avg 14177
Sorting by length 100%
Counting unique k-mers 100%
Clustering 100%
Sorting clusters 100%
Writing clusters 100%
Clusters: 24 Size min 1, max 1, avg 1.0
Singletons: 24, 100.0% of seqs, 100.0% of clusters
Sorting clusters by abundance 100%
vsearch v1.11.1_osx_x86_64, 32.0GB RAM, 8 cores
https://github.com/torognes/vsearch

CLUSTpred output has been stored in: Hsapiens_ChrY_ltrpred
Join Cluster table: nrow(df) = 24 candidates.
unique(ID) = 24 candidates.
unique(orf.id) = 24 candidates.
Join Cluster Copy Number table: nrow(df) = 24 candidates.
unique(ID) = 24 candidates.
unique(orf.id)) = 24 candidates.
Perform methylation context quantification..
Join methylation context (CG, CHG, CHH, CCG) count table: nrow(df) = 24 candidates.
unique(ID) = 24 candidates.
unique(orf.id) = 24 candidates.
Copy files to result folder 'Hsapiens_ChrY_ltrpred'.
The LTRpred prediction table has been filtered (default) to remove potential false positives. Predicted LTRs must have an PBS or Protein Domain and must fulfill thresholds: sim = 70%; #orfs = 0. Furthermore, TEs having more than 10% of N's in their sequence have also been removed.
Input #TEs: 24
Output #TEs: 14
Perform solo LTR Copy Number Estimation....
Run makeblastdb of the genome assembly...


Building a new DB, current time: 11/01/2017 14:15:19
New DB name:   /Library/Frameworks/R.framework/Versions/3.3/Resources/library/LTRpred/Hsapiens_ChrY.fa
New DB title:  /Library/Frameworks/R.framework/Versions/3.3/Resources/library/LTRpred/Hsapiens_ChrY.fa
Sequence type: Nucleotide
Keep MBits: T
Maximum file size: 1000000000B
Adding sequences from FASTA; added 1 sequences in 0.0813279 seconds.
Perform BLAST searches of 3' prime LTRs against genome assembly...
Perform BLAST searches of 5' prime LTRs against genome assembly...
Import BLAST results...
Filter hit results...
Estimate CNV for each LTR sequence...
Finished LTR CNV estimation!
LTRpred analysis finished properly.
Warning message:
The LTR copy number estimation returned an empty file. This suggests that there were no solo LTRs found in the input genome sequence.
```

The `LTRpred()` function internally generates a folder named `*_ltrpred` which
stores all output annotation and sequence files.


In detail, the following files and folders are generated by the `LTRpred()` function:

- `*_BetweenLTRSeqs.fsa` : DNA sequences of the region between the LTRs in fasta format.

- `*_Details.tsv` : A spread sheet containing detailed information about the predicted LTRs.

- `*_FullLTRRetrotransposonSeqs.fsa` : DNA sequences of the entire predicted LTR retrotransposon.

- `*_index.fsa` : The suffixarray index file used to predict putative LTR retrotransposons.

- `*_Prediction.gff` : A spread sheet containing detailed additional information about the predicted LTRs (partially redundant with the *_Details.tsv file).

- `*_index_ltrdigest.fsa` : The suffixarray index file used to predict putative LTR retrotransposonswith LTRdigest.

- `*_LTRdigestPrediction.gff` : A spread sheet containing detailed information about the predicted LTRs.

- `*-ltrdigest_tabout.csv` : A spread sheet containing additional detailed information about the predicted LTRs.

- `*-ltrdigest_complete.fas` : The full length DNA sequences of all predicted LTR transposons.

- `*-ltrdigest_conditions.csv` : Contains information about the parameters used for a given LTRdigest run.

- `*-ltrdigest_pbs.fas` : Stores the predicted PBS sequences for the putative LTR retrotransposons.

- `*-ltrdigest_ppt.fas` : Stores the predicted PPT sequences for the putative LTR retrotransposons.

- `*-ltrdigest_5ltr.fas` and `*-ltrdigest_3ltr.fas`: Stores the predicted 5' and 3' LTR sequences. Note: If the direction of the putative retrotransposon could be predicted, these files will contain the corresponding 3' and 5' LTR sequences. If no direction could be predicted, forward direction with regard to the original sequence will be assumed by LTRdigest, i.e. the 'left' LTR will be considered the 5' LTR.

- `*-ltrdigest_pdom_<domainname>.fas` : Stores the DNA sequences of the HMM matches to the LTR retrotransposon candidates.

- `*-ltrdigest_pdom_<domainname>_aa.fas` : Stores the concatenated protein sequences of the HMM matches to the LTR retrotransposon candidates.

- `*-ltrdigest_pdom_<domainname>_ali.fas` : Stores the alignment information for all matches of the given protein domain model to the translations of all candidates.

- `*_orfs_nt.fsa` : Stores the predicted open reading frames within the predicted LTR transposons as DNA sequence.

- `*_orfs_aa.fsa` : Stores the predicted open reading frames within the predicted LTR transposons as protein sequence.

- `*_LTRpred.gff` : Stores the LTRpred predicted LTR transposons in GFF format.

- `*_LTRpred.bed` : Stores the LTRpred predicted LTR transposons in BED format.

- `*_LTRpred_DataSheet.csv` : Stores the output table as data sheet.

The `LTRpred()` output table `*_LTRpred_DataSheet.csv` is in [tidy](http://vita.had.co.nz/papers/tidy-data.html) format and can then be imported using `read.ltrpred()`.
The `tidy` output format is designed to work seamlessly with the [tidyverse](https://www.tidyverse.org/) and [R data science](http://r4ds.had.co.nz/) framework.

```r
# import LTRpred prediction output
Hsapiens_chrY <- read.ltrpred("Hsapiens_ChrY_ltrpred/Hsapiens_ChrY_LTRpred_DataSheet.tsv")
# look at some results
dplyr::select(Hsapiens_chrY, ltr_similarity:end, tRNA_motif, Clust_cn)
```

Looking at all columns:

```r
dplyr::glimpse(Hsapiens_chrY)
```

```
Observations: 14
Variables: 91
$ species                 <chr> "Hsapiens_ChrY", "Hsapiens_ChrY", "Hsapiens_ChrY", "Hsapiens_ChrY", "Hsapiens_...
$ ID                      <chr> "Hsapiens_ChrY_LTR_retrotransposon1", "Hsapiens_ChrY_LTR_retrotransposon3", "H...
$ dfam_target_name        <chr> NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA
$ ltr_similarity          <dbl> 80.73, 89.85, 79.71, 84.93, 75.52, 76.47, 80.28, 76.92, 89.53, 93.57, 82.35, 7...
$ similarity              <chr> "(80,82]", "(88,90]", "(78,80]", "(84,86]", NA, "[76,78]", "(80,82]", "[76,78]...
$ protein_domain          <chr> "RVT_1", "RVT_1", NA, "RVT_1", "RVT_1", "RVT_1", NA, "RVT_1", "RVT_1", "RVT_1"...
$ orfs                    <int> 1, 1, 0, 0, 0, 1, 0, 0, 0, 1, 2, 0, 3, 1
$ chromosome              <chr> "NC000024.10Homosa", "NC000024.10Homosa", "NC000024.10Homosa", "NC000024.10Hom...
$ start                   <int> 3143582, 3275798, 3313536, 3390751, 3541388, 3568176, 3674859, 3918792, 432818...
$ end                     <int> 3162877, 3299928, 3318551, 3411614, 3547227, 3591658, 3695232, 3942018, 433468...
$ strand                  <chr> "-", "-", "+", "-", "-", "-", "-", "-", "+", "-", "+", "-", "+", "-"
$ width                   <int> 19296, 24131, 5016, 20864, 5840, 23483, 20374, 23227, 6498, 5840, 5724, 16759,...
$ annotation              <chr> "LTR_retrotransposon", "LTR_retrotransposon", "LTR_retrotransposon", "LTR_retr...
$ pred_tool               <chr> "LTRpred", "LTRpred", "LTRpred", "LTRpred", "LTRpred", "LTRpred", "LTRpred", "...
$ frame                   <chr> ".", ".", ".", ".", ".", ".", ".", ".", ".", ".", ".", ".", ".", "."
$ score                   <chr> ".", ".", ".", ".", ".", ".", ".", ".", ".", ".", ".", ".", ".", "."
$ lLTR_start              <int> 3143582, 3275798, 3313536, 3390751, 3541388, 3568176, 3674859, 3918792, 432818...
$ lLTR_end                <int> 3143687, 3276408, 3313665, 3390968, 3541530, 3568306, 3675142, 3918924, 432836...
$ lLTR_length             <int> 106, 611, 130, 218, 143, 131, 284, 133, 182, 451, 363, 122, 133, 455
$ rLTR_start              <int> 3162769, 3299338, 3318414, 3411396, 3547099, 3591523, 3694963, 3941876, 433449...
$ rLTR_end                <int> 3162877, 3299928, 3318551, 3411614, 3547227, 3591658, 3695232, 3942018, 433468...
$ rLTR_length             <int> 109, 591, 138, 219, 129, 136, 270, 143, 191, 450, 374, 120, 139, 453
$ lTSD_start              <int> 3143578, 3275794, 3313532, 3390747, 3541384, 3568172, 3674855, 3918788, 432818...
$ lTSD_end                <int> 3143581, 3275797, 3313535, 3390750, 3541387, 3568175, 3674858, 3918791, 432818...
$ lTSD_motif              <chr> "acag", "ttgt", "ttag", "ccca", "aggc", "ccct", "tttt", "aaat", "aaat", "gataa...
$ rTSD_start              <int> 3162878, 3299929, 3318552, 3411615, 3547228, 3591659, 3695233, 3942019, 433468...
$ rTSD_end                <int> 3162881, 3299932, 3318555, 3411618, 3547231, 3591662, 3695236, 3942022, 433468...
$ rTSD_motif              <chr> "acag", "ttgt", "ttag", "ccca", "aggc", "ccct", "tttt", "aaat", "aaat", "gataa...
$ PPT_start               <int> NA, NA, NA, NA, NA, NA, NA, NA, 4334511, NA, NA, NA, NA, NA
$ PPT_end                 <int> NA, NA, NA, NA, NA, NA, NA, NA, 4334521, NA, NA, NA, NA, NA
$ PPT_motif               <chr> NA, NA, NA, NA, NA, NA, NA, NA, "ggggaaaacaa", NA, NA, NA, NA, NA
$ PPT_strand              <chr> NA, NA, NA, NA, NA, NA, NA, NA, "+", NA, NA, NA, NA, NA
$ PPT_offset              <int> NA, NA, NA, NA, NA, NA, NA, NA, 27, NA, NA, NA, NA, NA
$ PBS_start               <int> NA, NA, 3313667, NA, NA, NA, 3694954, NA, NA, NA, NA, NA, NA, NA
$ PBS_end                 <int> NA, NA, 3313677, NA, NA, NA, 3694965, NA, NA, NA, NA, NA, NA, NA
$ PBS_strand              <chr> NA, NA, "+", NA, NA, NA, "-", NA, NA, NA, NA, NA, NA, NA
$ tRNA                    <chr> NA, NA, "Homo_sapiens_tRNA-Ser-AGA-6-1 (chr11.trna9-SerAGA) chr11:109165314-10...
$ tRNA_motif              <chr> NA, NA, "aattagctgga", NA, NA, NA, "agtgctgggatt", NA, NA, NA, NA, NA, NA, NA
$ PBS_offset              <int> NA, NA, 1, NA, NA, NA, 3, NA, NA, NA, NA, NA, NA, NA
$ tRNA_offset             <int> NA, NA, 1, NA, NA, NA, 4, NA, NA, NA, NA, NA, NA, NA
$ PBS/tRNA_edist          <int> NA, NA, 1, NA, NA, NA, 1, NA, NA, NA, NA, NA, NA, NA
$ orf.id                  <chr> "NC_000024.10_Homo_sa_3143582_3162877", "NC_000024.10_Homo_sa_3275798_3299928"...
$ repeat_region_length    <int> 19304, 24139, 5024, 20872, 5848, 23491, 20382, 23235, 6506, 5854, 5732, 16767,...
$ PPT_length              <int> NA, NA, NA, NA, NA, NA, NA, NA, 11, NA, NA, NA, NA, NA
$ PBS_length              <int> NA, NA, 11, NA, NA, NA, 12, NA, NA, NA, NA, NA, NA, NA
$ dfam_acc                <chr> NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA
$ dfam_bits               <dbl> NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA
$ dfam_e_value            <dbl> NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA
$ dfam_bias               <dbl> NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA
$ dfam_hmm-st             <dbl> NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA
$ dfam_hmm-en             <dbl> NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA
$ dfam_strand             <chr> NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA
$ dfam_ali-st             <dbl> NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA
$ dfam_ali-en             <dbl> NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA
$ dfam_env-st             <dbl> NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA
$ dfam_env-en             <dbl> NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA
$ dfam_modlen             <dbl> NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA
$ dfam_target_description <chr> NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA
$ Clust_Cluster           <chr> "unique", "unique", "unique", "unique", "unique", "unique", "unique", "unique"...
$ Clust_Target            <chr> "none", "none", "none", "none", "none", "none", "none", "none", "none", "none"...
$ Clust_Perc_Ident        <dbl> NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA
$ Clust_cn                <int> 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1
$ TE_CG_abs               <dbl> 62, 125, 35, 139, 26, 257, 176, 144, 15, 61, 32, 64, 160, 104
$ TE_CG_rel               <dbl> 0.003213101, 0.005180059, 0.006977671, 0.006662193, 0.004452055, 0.010944087, ...
$ TE_CHG_abs              <dbl> 659, 830, 150, 742, 198, 825, 780, 731, 173, 290, 258, 569, 742, 282
$ TE_CHG_rel              <dbl> 0.03415216, 0.03439559, 0.02990431, 0.03556365, 0.03390411, 0.03513180, 0.0382...
$ TE_CHH_abs              <dbl> 2571, 3454, 748, 3150, 836, 3231, 2854, 3078, 905, 1530, 940, 2316, 2857, 1461
$ TE_CHH_rel              <dbl> 0.1332400, 0.1431354, 0.1491228, 0.1509778, 0.1431507, 0.1375889, 0.1400805, 0...
$ TE_CCG_abs              <dbl> 13, 24, 6, 33, 4, 69, 52, 26, 1, 19, 9, 14, 41, 40
$ TE_CCG_rel              <dbl> 0.0006737148, 0.0009945713, 0.0011961722, 0.0015816718, 0.0006849315, 0.002938...
$ TE_N_abs                <dbl> 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0
$ CG_3ltr_abs             <dbl> 1, 0, 1, 8, 1, 11, 11, 0, 0, 7, 0, 1, 9, 9
$ CG_3ltr_rel             <dbl> 0.009433962, 0.000000000, 0.007246377, 0.036697248, 0.006993007, 0.083969466, ...
$ CHG_3ltr_abs            <dbl> 2, 24, 9, 14, 9, 9, 18, 4, 9, 22, 13, 5, 8, 25
$ CHG_3ltr_rel            <dbl> 0.01886792, 0.03927987, 0.06521739, 0.06422018, 0.06293706, 0.06870229, 0.0633...
$ CHH_3ltr_abs            <dbl> 18, 69, 18, 43, 23, 13, 54, 24, 26, 120, 62, 18, 25, 113
$ CHH_3ltr_rel            <dbl> 0.16981132, 0.11292962, 0.13043478, 0.19724771, 0.16083916, 0.09923664, 0.1901...
$ CCG_3ltr_abs            <dbl> 0, 0, 0, 2, 1, 4, 5, 0, 0, 1, 0, 0, 4, 1
$ CCG_3ltr_rel            <dbl> 0.000000000, 0.000000000, 0.000000000, 0.009174312, 0.006993007, 0.030534351, ...
$ N_3ltr_abs              <dbl> 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0
$ CG_5ltr_abs             <dbl> 1, 0, 1, 8, 1, 11, 11, 0, 0, 7, 0, 1, 9, 9
$ CG_5ltr_rel             <dbl> 0.009433962, 0.000000000, 0.007246377, 0.036697248, 0.006993007, 0.083969466, ...
$ CHG_5ltr_abs            <dbl> 2, 24, 9, 14, 9, 9, 18, 4, 9, 22, 13, 5, 8, 25
$ CHG_5ltr_rel            <dbl> 0.01886792, 0.03927987, 0.06521739, 0.06422018, 0.06293706, 0.06870229, 0.0633...
$ CHH_5ltr_abs            <dbl> 18, 69, 18, 43, 23, 13, 54, 24, 26, 120, 62, 18, 25, 113
$ CHH_5ltr_rel            <dbl> 0.16981132, 0.11292962, 0.13043478, 0.19724771, 0.16083916, 0.09923664, 0.1901...
$ CCG_5ltr_abs            <dbl> 0, 0, 0, 2, 1, 4, 5, 0, 0, 1, 0, 0, 4, 1
$ CCG_5ltr_rel            <dbl> 0.000000000, 0.000000000, 0.000000000, 0.009174312, 0.006993007, 0.030534351, ...
$ N_5ltr_abs              <dbl> 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0
$ cn_3ltr                 <dbl> NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA
$ cn_5ltr                 <dbl> NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA
```

### Output file format of `LTRpred()`

The columns of the prediction file store the following information:

tidy format (each row contains all information for a predicted LTR retrotransposon):

- `species` : Name of the species or genomic sequence that is passed as input to `LTRpred`
- `ID` : unique identifier that is generated by `LTRpred` to mark individual LTR retrotransposon predictions
- `dfam_target_name` : 
- `ltr_similarity` : sequence similarity (global alignment) between the 5' and 3' long terminal repeat (LTR)
- `similarity`: `ltr_similarity` classified into 2% intervals
- `chromosome` : chromosome at which LTR retrotransposon was found
- `start` : start coordinate of the LTR retrotransposon within the chromosome
- `end` : end coordinate of the LTR retrotransposon within the chromosome
- `strand` : strand on which the corresponding LTR retrotransposon was found
- `dfam_acc` :
- `width` : length/width of the LTR retrotransposon 
- `orfs` : number of predicted open reading frames within the LTR retrotransposon sequence
- `protein_domain` : name of protein domains found between the 5' and 3' LTRs
- `Clust_Cluster` :
- `Clust_Target` :
- `Clust_Perc_Ident` : sequence similarity (global alignment) with `Clust_Target` 
- `Clust_cn` : number of LTR retrotransposon that belong to the same cluster (see `Clust_Target`) - if `Clust_cn` = 1 -> unique element
- `lLTR_start` : start coordinate of the 5' LTR within the chromosome
- `lLTR_end` : end coordinate of the 5' LTR within the chromosome
- `lLTR_length` : length/width of the 5' LTR 
- `rLTR_start` : start coordinate of the 3' LTR within the chromosome
- `rLTR_end` : end coordinate of the 3' LTR within the chromosome
- `rLTR_length` : length/width of the 3' LTR
- `lTSD_start` : start coordinate of the 5' target site duplication (TSD) within the chromosome
- `lTSD_end` : end coordinate of the 5' target site duplication (TSD) within the chromosome
- `lTSD_motif` : 5' TSD sequence motif
- `PPT_start` : start coordinate of the PPT motif within the chromosome
- `PPT_end` : end coordinate of the PPT motif within the chromosome
- `PPT_motif` : PPT sequence motif
- `PPT_strand` : strand of PPT sequence 
- `PPT_offset` : width between `PPT_end` and `rLTR_start`
- `PBS_start` : start coordinate of the PBS motif within the chromosome
- `PBS_end` : end coordinate of the PBS motif within the chromosome
- `PBS_strand` : strand of PBS sequence
- `tRNA` : name of tRNA that matches PBS sequence
- `tRNA_motif` : tRNA sequence motif
- `PBS_offset` : width between `lLTR_end` and `PBS_start`
- `tRNA_offset` :
- `PBS/tRNA_edist` : edit distance between tRNA motif and PBS motif
- `PPT_length` : length/width of the PPT motif
- `PBS_length` : length/width of the PBS motif
- `dfam_target_description` : description of Dfam target hit
- `orf.id` : id generated by ORF prediction
- `TE_CG_abs` : total number of CG motifs found within the full length LTR retrotransposon sequence
- `TE_CG_rel` : relative number of CG motifs found within the full length LTR retrotransposon (normalized by element length - `width`)
- `TE_CHG_abs` : total number of CHG motifs found within the full length LTR retrotransposon sequence
- `TE_CHG_rel` : relative number of CHG motifs found within the full length LTR retrotransposon sequence (normalized by element length - `width`)
- `TE_CHH_abs` : total number of CHH motifs found within the full length LTR retrotransposon sequence
- `TE_CHH_rel` : relative number of CHH motifs found within the full length LTR retrotransposon sequence (normalized by element length - `width`)
- `TE_N_abs` : total number of N's found within the full length LTR retrotransposon sequence
- `CG_3ltr_abs` : total number of CG motifs found within only the 3' LTR sequence
- `CG_3ltr_rel` : relative number of CG motifs found within only the 3' LTR sequence (normalized by LTR length)
- `CHG_3ltr_abs` : total number of CHG motifs found within only the 3' LTR sequence
- `CHG_3ltr_rel` : relative number of CHG motifs found within only the 3' LTR sequence (normalized by LTR length)
- `CHH_3ltr_abs` : total number of CHH motifs found within only the 3' LTR sequence
- `CG_5ltr_abs` : total number of CG motifs found within only the 5' LTR sequence
- `CG_5ltr_rel` : relative number of CG motifs found within only the 5' LTR sequence (normalized by LTR length)
- `N_3ltr_abs` : total number of N's found within only the 3' LTR sequence
- `CHG_5ltr_abs` : total number of CHG motifs found within only the 5' LTR sequence
- `CHG_5ltr_rel` : relative number of CHG motifs found within only the 5' LTR sequence (normalized by LTR length)
- `CHH_5ltr_abs` : total number of CHH motifs found within only the 5' LTR sequence
- `CHH_5ltr_rel` : relative number of CHH motifs found within only the 5' LTR sequence (normalized by LTR length)
- `N_5ltr_abs` : total number of N's found within only the 5' LTR sequence
- `cn_3ltr` : number of solo 3' LTRs found in the genome
- `cn_5ltr` : number of solo 5' LTRs found in the genome (minus number of solo 3' LTRs)
# Detailed LTRpred run
