<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>De novo functional annotation of retrotransposons  • LTRpred</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="De novo functional annotation of retrotransposons ">
<meta property="og:description" content="A de novo functional annotation pipeline to predict potentially active LTR retrotransposons in any genome assembly file.">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">LTRpred</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="articles/Introduction.html">Introduction to LTRpred</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/HajkD/LTRpred/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">

<div id="ltrpredict-de-novo-annotation-of-intact-retrotransposons-" class="section level2">
<h2 class="hasAnchor">
<a href="#ltrpredict-de-novo-annotation-of-intact-retrotransposons-" class="anchor"></a><strong>LTRpred(ict)</strong>: <em>de novo</em> annotation of intact retrotransposons <img src="inst/LTRpred_logo.png" align="right" height="174" width="150">
</h2>

<p>Transposable elements (TEs) comprise vast parts of eukaryotic genomes. In the past, TEs were seen as selfish mobile elements capable of populating a host genome to increase their chances for survival. By doing so they leave traces of junk DNA in host genomes that are usually regarded as by-products when sequencing, assembling, and annotating new genomes.</p>
<p>However, this picture is slowly changing (<a href="https://academic.oup.com/gbe/article/11/12/3382/5637757">Drost &amp; Sanchez, 2019</a>) and TEs have been shown to be involved in generating a diverse range of novel phenotypes.</p>
<p>Today, the <em>de novo</em> detection of transposable elements is performed by annotation tools which try to detect any type of repeated sequence, TE family, or remnand DNA loci that can be associated with a known transposable element within a genome assembly. The main goal of such efforts is to retrieve a maximum amount of loci that can be associated with TEs. If successful, such annotation can then be used to mask host genomes and to perform classic (phylo-)genomics studies focusing on host genes.</p>
<p>More than <a href="https://docs.google.com/spreadsheets/d/1UBK70zExiL0gFVaIAILiGhflCGXAq_SF_lymaxTE1pY/edit#gid=0">300 repeat and TE annotation tools</a> have been developed so far. Most of them are designed and optimized to annotate either the entire repeat space or specific superfamilies of TEs and their DNA remnants.</p>
<blockquote>
<p>The LTRpred pipeline has a different goal than all other annotation tools. It focuses particularly on <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC463057/">LTR retrotransposons</a> and aims to annotate only functional and potentially mobile elements. Such type of annotation is crucial for studying retrotransposon activity in eukaryotic genomes and to understand whether specific retrotransposon families can be activated artificially and harnessed to <a href="https://www.slcu.cam.ac.uk/news/tomato-jumping-genes">mutagenize genomes at much faster speed</a>.</p>
</blockquote>
<p>In detail, <code>LTRpred</code> will take any genome assembly file in <code>fasta</code> format as input and will generate a detailed annotation of functional and potentially mobile LTR retrotransposons.</p>
<p><img src="articles/LTRfeatures.png"></p>
<p><strong>Users can consult a comprehensive <a href="https://hajkd.github.io/LTRpred/articles/Introduction.html">Introduction</a> to the <code>LTRpred</code> pipeline to get familiar with the tool.</strong></p>
</div>
<div id="install" class="section level2">
<h2 class="hasAnchor">
<a href="#install" class="anchor"></a>Install</h2>
<p>The fastest way to install <code>LTRpred</code> is via a <a href="https://hub.docker.com/r/drostlab/ltrpred">Docker container</a>. Please make sure to read the <a href="https://hajkd.github.io/LTRpred/articles/Introduction.html#installation">detailed installation instructions</a> to be able to pass data to the container.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1"></a><span class="co"># retrieve docker image from dockerhub</span></span>
<span id="cb1-2"><a href="#cb1-2"></a><span class="ex">docker</span> pull drostlab/ltrpred</span>
<span id="cb1-3"><a href="#cb1-3"></a><span class="co"># run ltrpred container</span></span>
<span id="cb1-4"><a href="#cb1-4"></a><span class="ex">docker</span> run --rm -ti drostlab/ltrpred</span>
<span id="cb1-5"><a href="#cb1-5"></a><span class="co"># start R prompt within ltrpred container</span></span>
<span id="cb1-6"><a href="#cb1-6"></a><span class="ex">~</span>:/<span class="ex">app</span># R</span></code></pre></div>
<p>Users who wish to run the <code>LTRpred</code> Docker container in a <a href="https://docs.conda.io/en/latest/">conda</a> environment can use the <a href="https://github.com/HajkD/LTRpred/issues/16">following approach based on UDocker</a> (Many thanks to Ilja Bezrukov).</p>
</div>
<div id="citation" class="section level2">
<h2 class="hasAnchor">
<a href="#citation" class="anchor"></a>Citation</h2>
<p>Please cite the following paper when using <code>LTRpred</code> for your own research:</p>
<blockquote>
<p>HG Drost. <a href="https://joss.theoj.org/papers/10.21105/joss.02170"><strong>LTRpred: <em>de novo</em> annotation of intact retrotransposons</strong></a>. <strong>Journal of Open Source Software</strong>, 5(50), 2170 (2020).</p>
</blockquote>
</div>
<div id="tutorials" class="section level2">
<h2 class="hasAnchor">
<a href="#tutorials" class="anchor"></a>Tutorials</h2>
<div id="quick-start" class="section level3">
<h3 class="hasAnchor">
<a href="#quick-start" class="anchor"></a>Quick Start</h3>
<p>The fastest way to generate a LTR retrotransposon prediction for a genome of interest (after <a href="https://hajkd.github.io/LTRpred/articles/Introduction.html">installing</a> all prerequisite command line tools) is to use the <code><a href="reference/LTRpred.html">LTRpred()</a></code> function and relying on the default parameters. In the following example, a LTR transposon prediction is performed for parts of the Human Y chromosome.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Perform de novo LTR transposon prediction for the Human Y chromosome</span>
<span class="fu">LTRpred</span><span class="fu">::</span><span class="fu"><a href="reference/LTRpred.html">LTRpred</a></span><span class="op">(</span>genome.file <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"Hsapiens_ChrY.fa"</span>, package <span class="op">=</span> <span class="st">"LTRpred"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p>When running your own genome, please specify <code>genome.file = "path/to/your/genome.fasta</code> instead of <code><a href="https://rdrr.io/r/base/system.file.html">system.file(..., package = "LTRpred")</a></code>. The command <code><a href="https://rdrr.io/r/base/system.file.html">system.file(..., package = "LTRpred")</a></code> merely references the path to the example file stored in the LTRpred package itself.</p>
<p>This tutorial introduces users to <code>LTRpred</code>:</p>
<ul>
<li><a href="https://hajkd.github.io/LTRpred/articles/Introduction.html">Introduction to LTRpred</a></li>
</ul>
<p>Users can also read the tutorials within (<a href="http://www.rstudio.com/">RStudio</a>) :</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/HajkD/LTRpred">LTRpred</a></span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/browseVignettes.html">browseVignettes</a></span><span class="op">(</span><span class="st">"LTRpred"</span><span class="op">)</span></code></pre></div>
<p>You can also find a list of all available <code>LTRpred</code> functions here: <a href="https://hajkd.github.io/LTRpred/reference/index.html" class="uri">https://hajkd.github.io/LTRpred/reference/index.html</a></p>
</div>
<div id="studies-that-successfully-used-ltrpred-to-annotate-functional-retrotransposons" class="section level3">
<h3 class="hasAnchor">
<a href="#studies-that-successfully-used-ltrpred-to-annotate-functional-retrotransposons" class="anchor"></a>Studies that successfully used <code>LTRpred</code> to annotate functional retrotransposons</h3>
<blockquote>
<ul>
<li><p>Z Wang &amp; D Baulcombe. <a href="https://www.nature.com/articles/s41467-020-14995-6"><strong>Transposon age and non-CG methylation</strong></a>. <strong>Nature Communications</strong>, 11, 1221 (2020).</p></li>
<li><p>JH Collins, KW Keating, TR Jones, S Balaji, CB Marsan et al. <a href="https://www.nature.com/articles/s41467-021-21656-9"><strong>Engineered yeast genomes accurately assembled from pure and mixed samples</strong></a>. <strong>Nature communications</strong>, 12, 1485 (2021).</p></li>
<li><p>H Kundariya et al. <a href="https://www.nature.com/articles/s41467-020-19140-x"><strong>MSH1-induced heritable enhanced growth vigor through grafting is associated with the RdDM pathway in plants</strong></a> <strong>Nature Communications</strong>, 11, 5343 (2020).</p></li>
<li><p>J Cho, M Benoit, M Catoni, <strong>HG Drost</strong>, A Brestovitsky, M Oosterbeek and J Paszkowski. <a href="https://www.nature.com/articles/s41477-018-0320-9"><strong>Sensitive detection of pre-integration intermediates of LTR retrotransposons in crop plants</strong></a>. <strong>Nature Plants</strong>, 5, 26-33 (2019).</p></li>
<li><p>M Benoit, <strong>HG Drost</strong>, M Catoni, Q Gouil, S Lopez-Gomollon, DC Baulcombe, J Paszkowski. <a href="https://journals.plos.org/plosgenetics/article?id=10.1371/journal.pgen.1008370"><strong>Environmental and epigenetic regulation of Rider retrotransposons in tomato</strong></a>. <strong>PloS Genetics</strong>, 15(9): e1008370 (2019).</p></li>
<li><p>Nguinkal <em>et al.</em> <a href="https://www.mdpi.com/2073-4425/10/9/708/htm"><strong>The First Highly Contiguous Genome Assembly of Pikeperch (Sander lucioperca), an Emerging Aquaculture Species in Europe</strong></a> <strong>Genes</strong>, 0(9), 708 (2019).</p></li>
<li><p>E Cerruti, C Gisbert, <strong>HG Drost</strong>, D Valentino, E Portis, L Barchi, J Prohens, S Lanteri, C Comino, M Catoni. <a href="https://www.biorxiv.org/content/10.1101/831719v1"><strong>Epigenetic bases of grafting-induced vigour in eggplant</strong></a>. <strong>bioaRxiv</strong> (2019).</p></li>
<li><p>P Gan, R Hiroyama, A Tsushima, S Masuda <em>et al</em>. <a href="https://www.biorxiv.org/content/10.1101/2020.04.28.061093v1.abstract"><strong>Subtelomeric regions and a repeat-rich chromosome harbor multicopy effector gene clusters with variable conservation in multiple plant pathogenic Colletotrichum species</strong></a> <strong>bioRxiv</strong> (2020)</p></li>
<li><p>J Wang et al. <a href="https://www.biorxiv.org/content/10.1101/2020.08.19.257527v1"><strong>Gigantic Genomes Can Provide Empirical Tests of TE Dynamics Models–An Example from Amphibians</strong></a>. <strong>bioRxiv</strong>, (2020).</p></li>
<li><p>Y Ayukawa et al. <a href="https://www.biorxiv.org/content/10.1101/2020.10.06.329052v1"><strong>A pair of effectors encoded on a conditionally dispensable chromosome of Fusarium oxysporum suppress host-specific immunity</strong></a>. <strong>bioRxiv</strong>, (2020).</p></li>
</ul>
<p>C Meguerditchian, A Ergun, V Decroocq, M Lefebvre et al. <a href="https://www.biorxiv.org/content/10.1101/2021.02.25.432867v1.abstract"><strong>Pipeline to detect the relationship between transposable elements and adjacent genes in host genome</strong></a> <strong>bioRxiv</strong>, (2021).</p>
</blockquote>
</div>
</div>
<div id="discussions-and-bug-reports" class="section level2">
<h2 class="hasAnchor">
<a href="#discussions-and-bug-reports" class="anchor"></a>Discussions and Bug Reports</h2>
<p>I would be very happy to learn more about potential improvements of the concepts and functions provided in this package.</p>
<p>Furthermore, in case you find some bugs or need additional (more flexible) functionality of parts of this package, please let me know:</p>
<p><a href="https://github.com/HajkD/LTRpred/issues" class="uri">https://github.com/HajkD/LTRpred/issues</a></p>
</div>
<div id="acknowledgement" class="section level2">
<h2 class="hasAnchor">
<a href="#acknowledgement" class="anchor"></a>Acknowledgement</h2>
<p>I would like to thank the <a href="http://www.slcu.cam.ac.uk/research/paszkowski-group/group-members">Paszkowski team</a> for incredible support and motivating discussions that led to the realization of this project.</p>
</div>

  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="links">
<h2>Links</h2>
<ul class="list-unstyled">
<li>Browse source code at <br><a href="https://github.com/HajkD/LTRpred/">https://​github.com/​HajkD/​LTRpred/​</a>
</li>
<li>Report a bug at <br><a href="https://github.com/HajkD/LTRpred/issues">https://​github.com/​HajkD/​LTRpred/​issues</a>
</li>
</ul>
</div>
<div class="license">
<h2>License</h2>
<ul class="list-unstyled">
<li><a href="https://www.r-project.org/Licenses/GPL-2">GPL-2</a></li>
</ul>
</div>
<div class="developers">
<h2>Developers</h2>
<ul class="list-unstyled">
<li>Hajk-Georg Drost <br><small class="roles"> Author, maintainer </small>  </li>
</ul>
</div>

  <div class="dev-status">
<h2>Dev status</h2>
<ul class="list-unstyled">
<li><a href="https://joss.theoj.org/papers/eeb2359d2459d3ae448cafac3ae33211"><img src="https://joss.theoj.org/papers/eeb2359d2459d3ae448cafac3ae33211/status.svg" alt="status"></a></li>
</ul>
</div>
</div>
</div>


      <footer><div class="copyright">
  <p>Developed by Hajk-Georg Drost.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
