<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>De novo annotation of potentially active LTR retrotransposons  • LTRpred</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="De novo annotation of potentially active LTR retrotransposons ">
<meta property="og:description" content="A meta-genomics tool explicitly designed to investigate LTR retrotransposons in a comparative manner.">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">LTRpred</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.0.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="articles/Introduction.html">Introduction to LTRpred</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/HajkD/LTRpred">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">

<div id="ltrpredict-a-pipeline-for-automated-functional-annotation-of-ltr-retrotransposons-for-comparative-genomics-studies" class="section level2">
<h2 class="hasAnchor">
<a href="#ltrpredict-a-pipeline-for-automated-functional-annotation-of-ltr-retrotransposons-for-comparative-genomics-studies" class="anchor"></a>LTRpred(ict): a pipeline for automated functional annotation of LTR retrotransposons for comparative genomics studies <img src="inst/LTRpred_logo.png" align="right" height="174" width="150">
</h2>
<p>An easy way to perform <strong>de novo</strong> functional annotation of <code>LTR retrotransposons</code> from any genome assembly in <code>fasta</code> format.</p>
<p><img src="articles/LTRfeatures.png"></p>
<p>Users can consult a comprehensive <a href="https://hajkd.github.io/LTRpred/articles/Introduction.html">Introduction</a> to the <code>LTRpred</code> pipeline to get familiar with the tool.</p>
</div>
<div id="install" class="section level2">
<h2 class="hasAnchor">
<a href="#install" class="anchor"></a>Install</h2>
<p>Install prerequisite CRAN and Bioconductor packages:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" data-line-number="1"><span class="kw"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"tidyverse"</span>, <span class="st">"data.table"</span>, <span class="st">"seqinr"</span>, <span class="st">"biomartr"</span>, <span class="st">"dtplyr"</span>, <span class="st">"devtools"</span>))</a>
<a class="sourceLine" id="cb1-2" data-line-number="2"></a>
<a class="sourceLine" id="cb1-3" data-line-number="3"><span class="cf">if</span> (<span class="op">!</span><span class="kw"><a href="https://rdrr.io/r/base/ns-load.html">requireNamespace</a></span>(<span class="st">"BiocManager"</span>, <span class="dt">quietly =</span> <span class="ot">TRUE</span>))</a>
<a class="sourceLine" id="cb1-4" data-line-number="4">    <span class="kw"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span>(<span class="st">"BiocManager"</span>)</a>
<a class="sourceLine" id="cb1-5" data-line-number="5">BiocManager<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/BiocManager/man/install.html">install</a></span>()</a>
<a class="sourceLine" id="cb1-6" data-line-number="6"></a>
<a class="sourceLine" id="cb1-7" data-line-number="7">BiocManager<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/BiocManager/man/install.html">install</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"rtracklayer"</span>, <span class="st">"GenomicFeatures"</span>, <span class="st">"biomaRt"</span>))</a>
<a class="sourceLine" id="cb1-8" data-line-number="8"></a>
<a class="sourceLine" id="cb1-9" data-line-number="9">devtools<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/devtools/man/remote-reexports.html">install_github</a></span>(<span class="st">"HajkD/metablastr"</span>, <span class="dt">build_vignettes =</span> <span class="ot">TRUE</span>, <span class="dt">dependencies =</span> <span class="ot">TRUE</span>)</a>
<a class="sourceLine" id="cb1-10" data-line-number="10"></a>
<a class="sourceLine" id="cb1-11" data-line-number="11"><span class="kw"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"BSDA"</span>, <span class="st">"ggrepel"</span>, <span class="st">"gridExtra"</span>)) </a></code></pre></div>
<p>Now users may install <code>LTRpred</code> as follows:</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb2-1" data-line-number="1"><span class="co"># install.packages("devtools")</span></a>
<a class="sourceLine" id="cb2-2" data-line-number="2">devtools<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/devtools/man/remote-reexports.html">install_github</a></span>(<span class="st">"HajkD/LTRpred"</span>)</a></code></pre></div>
</div>
<div id="tutorials" class="section level2">
<h2 class="hasAnchor">
<a href="#tutorials" class="anchor"></a>Tutorials</h2>
<div id="quick-start" class="section level3">
<h3 class="hasAnchor">
<a href="#quick-start" class="anchor"></a>Quick Start</h3>
<p>The fastest way to generate a LTR retrotransposon prediction for a genome of interest (after <a href="https://hajkd.github.io/LTRpred/articles/Introduction.html">installing</a> all prerequisite command line tools) is to use the <code><a href="reference/LTRpred.html">LTRpred()</a></code> function and relying on the default parameters. In the following example, a LTR transposon prediction is performed for parts of the Human Y chromosome.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb3-1" data-line-number="1"><span class="co"># load LTRpred package</span></a>
<a class="sourceLine" id="cb3-2" data-line-number="2"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(LTRpred)</a>
<a class="sourceLine" id="cb3-3" data-line-number="3"><span class="co"># de novo LTR transposon prediction for the Human Y chromosome</span></a>
<a class="sourceLine" id="cb3-4" data-line-number="4"><span class="kw"><a href="reference/LTRpred.html">LTRpred</a></span>(<span class="dt">genome.file =</span> <span class="kw"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="st">"Hsapiens_ChrY.fa"</span>, <span class="dt">package =</span> <span class="st">"LTRpred"</span>))</a></code></pre></div>
<p>When running your own genome, please specify <code>genome.file = "path/to/your/genome.fasta</code> instead of <code><a href="https://rdrr.io/r/base/system.file.html">system.file(..., package = "LTRpred")</a></code>. The command <code><a href="https://rdrr.io/r/base/system.file.html">system.file(..., package = "LTRpred")</a></code> merely references the path to the example file stored in the LTRpred package itself.</p>
</div>
</div>
<div id="citation" class="section level2">
<h2 class="hasAnchor">
<a href="#citation" class="anchor"></a>Citation</h2>
<p>The <code>LTRpred</code> package is not formally published yet, but a manuscript is in preparation. For now, please cite one of the the following paper when using <code>LTRpred</code> for your own research. <code>LTRpred</code> is part of these studies and helped to predict potentially active retrotransposons that were later confirmed experimentally.</p>
<blockquote>
<p>M Benoit, <strong>HG Drost</strong>, M Catoni, Q Gouil, S Lopez-Gomollon, DC Baulcombe, J Paszkowski. <a href="https://journals.plos.org/plosgenetics/article?id=10.1371/journal.pgen.1008370"><strong>Environmental and epigenetic regulation of Rider retrotransposons in tomato</strong></a>. <em>PloS Genetics</em>, 15(9): e1008370 (2019).</p>
</blockquote>
<p>or</p>
<blockquote>
<p>J Cho, M Benoit, M Catoni, <strong>HG Drost</strong>, A Brestovitsky, M Oosterbeek and J Paszkowski. <a href="https://www.nature.com/articles/s41477-018-0320-9"><strong>Sensitive detection of pre-integration intermediates of LTR retrotransposons in crop plants</strong></a>. <em>Nature Plants</em>, 5, 26-33 (2019).</p>
</blockquote>
<p>This tutorial introduces users to <code>LTRpred</code>:</p>
<ul>
<li><a href="https://hajkd.github.io/LTRpred/articles/Introduction.html">Introduction to LTRpred</a></li>
</ul>
<p>Users can also read the tutorials within (<a href="http://www.rstudio.com/">RStudio</a>) :</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb4-1" data-line-number="1"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(LTRpred)</a>
<a class="sourceLine" id="cb4-2" data-line-number="2"><span class="kw"><a href="https://rdrr.io/r/utils/browseVignettes.html">browseVignettes</a></span>(<span class="st">"LTRpred"</span>)</a></code></pre></div>
<div id="studies-that-successfully-used-ltrpred-to-annotate-functional-retrotransposons" class="section level3">
<h3 class="hasAnchor">
<a href="#studies-that-successfully-used-ltrpred-to-annotate-functional-retrotransposons" class="anchor"></a>Studies that successfully used <code>LTRpred</code> to annotate functional retrotransposons</h3>
<blockquote>
<ul>
<li><p>J Cho, M Benoit, M Catoni, <strong>HG Drost</strong>, A Brestovitsky, M Oosterbeek and J Paszkowski. <a href="https://www.nature.com/articles/s41477-018-0320-9"><strong>Sensitive detection of pre-integration intermediates of LTR retrotransposons in crop plants</strong></a>. <strong>Nature Plants</strong>, 5, 26-33 (2019).</p></li>
<li><p>M Benoit, <strong>HG Drost</strong>, M Catoni, Q Gouil, S Lopez-Gomollon, DC Baulcombe, J Paszkowski. <a href="https://journals.plos.org/plosgenetics/article?id=10.1371/journal.pgen.1008370"><strong>Environmental and epigenetic regulation of Rider retrotransposons in tomato</strong></a>. <strong>PloS Genetics</strong> (2019) (<strong>in press</strong>).</p></li>
<li><p>E Cerruti, C Gisbert, <strong>HG Drost</strong>, D Valentino, E Portis, L Barchi, J Prohens, S Lanteri, C Comino, M Catoni. <a href="https://www.biorxiv.org/content/10.1101/831719v1"><strong>Epigenetic bases of grafting-induced vigour in eggplant</strong></a>. <strong>bioaRxiv</strong> (2019).</p></li>
<li><p>Nguinkal <em>et al.</em> <a href="https://www.mdpi.com/2073-4425/10/9/708/htm"><strong>The First Highly Contiguous Genome Assembly of Pikeperch (Sander lucioperca), an Emerging Aquaculture Species in Europe</strong></a> <strong>Genes</strong>, 0(9), 708 (2019).</p></li>
</ul>
</blockquote>
</div>
</div>
<div id="discussions-and-bug-reports" class="section level2">
<h2 class="hasAnchor">
<a href="#discussions-and-bug-reports" class="anchor"></a>Discussions and Bug Reports</h2>
<p>I would be very happy to learn more about potential improvements of the concepts and functions provided in this package.</p>
<p>Furthermore, in case you find some bugs or need additional (more flexible) functionality of parts of this package, please let me know:</p>
<p><a href="https://github.com/HajkD/LTRpred/issues" class="uri">https://github.com/HajkD/LTRpred/issues</a></p>
<p>In the <code>LTRpred</code> framework users can find:</p>
<div id="de-novo-prediction-and-annotation" class="section level3">
<h3 class="hasAnchor">
<a href="#de-novo-prediction-and-annotation" class="anchor"></a><em>De novo</em> prediction and annotation</h3>
<ul>
<li>
<em>de novo</em> prediction of LTR retrotransposons (nested, overlapping, or pure template) using <a href="http://www.zbh.uni-hamburg.de/forschung/arbeitsgruppe-genominformatik/software/ltrharvest.html">LTRharvest</a> and <a href="http://www.zbh.uni-hamburg.de/forschung/gi/software/ltrdigest.html">LTRdigest</a>
</li>
<li>annotation of predicted LTR retrotransposons using <a href="http://dfam.org/">Dfam</a> or <a href="http://www.girinst.org/repbase/">Repbase</a> as reference</li>
<li>solo LTR prediction based on specialized <a href="https://blast.ncbi.nlm.nih.gov/Blast.cgi?PAGE_TYPE=BlastDocs&amp;DOC_TYPE=Download">BLAST</a> searches</li>
<li>LTR retrotransposons family clustering using <a href="https://github.com/torognes/vsearch">vsearch</a>
</li>
<li>open reading frame prediction in LTR retrotransposons using <a href="https://www.drive5.com/usearch/">usearch</a>
</li>
<li>age estimation of predicted LTR retrotransposons in Mya (not implemented yet, but soon to come..)</li>
<li>CHH, CHG, CG, … content quantification in predicted LTR retrotransposons</li>
<li>filtering for (potentially) functional LTR retrotransposons<br>
</li>
<li>quality assesment of input genomes used to predict LTR retrotransposons</li>
</ul>
</div>
<div id="meta-genomics-analyses" class="section level3">
<h3 class="hasAnchor">
<a href="#meta-genomics-analyses" class="anchor"></a>Meta-Genomics Analyses</h3>
<ul>
<li>run <code>LTRpred</code> on entire kingdoms of life using only one command (see <code><a href="reference/LTRpred.meta.html">?LTRpred.meta</a></code>)</li>
<li>perform meta genomics studies customized for LTR retrotransposons</li>
<li>cluster LTR retrotransposons within and between species</li>
<li>quantify the diversity space of LTR retrotransposons for entire kingdoms of life</li>
</ul>
<div id="de-novo-annotation-functions" class="section level4">
<h4 class="hasAnchor">
<a href="#de-novo-annotation-functions" class="anchor"></a>De Novo Annotation Functions:</h4>
<ul>
<li>
<code><a href="reference/LTRpred.html">LTRpred()</a></code> : Major pipeline to predict LTR retrotransposons in a given genome</li>
<li>
<code>LTRpred.meta</code> : Perform Meta-Analyses with LTRpred</li>
<li>
<code><a href="reference/meta.summarize.html">meta.summarize()</a></code> : Summarize (concatenate) all predictions of a <code><a href="reference/LTRpred.meta.html">LTRpred.meta()</a></code> run</li>
<li>
<code><a href="reference/meta.apply.html">meta.apply()</a></code> : Apply functions to meta data generated by <code><a href="reference/LTRpred.html">LTRpred()</a></code>
</li>
<li>
<code><a href="reference/LTRharvest.html">LTRharvest()</a></code> : Run LTRharvest to predict putative LTR Retrotransposons</li>
<li>
<code><a href="reference/LTRdigest.html">LTRdigest()</a></code> : Run LTRdigest to predict putative LTR Retrotransposons</li>
</ul>
</div>
<div id="sequence-clustering-and-similarity-computations" class="section level4">
<h4 class="hasAnchor">
<a href="#sequence-clustering-and-similarity-computations" class="anchor"></a>Sequence Clustering and Similarity Computations</h4>
<ul>
<li>
<code><a href="reference/CLUSTpred.html">CLUSTpred()</a></code> : Cluster Sequences with VSEARCH</li>
<li>
<code><a href="reference/cluster.members.html">cluster.members()</a></code> : Select members of a specific cluster</li>
<li>
<code><a href="reference/clust2fasta.html">clust2fasta()</a></code> : Export sequences of TEs belonging to the same cluster to fasta files</li>
<li>
<code><a href="reference/AllPairwiseAlign.html">AllPairwiseAlign()</a></code> : Compute all pairwise (global) alignments with VSEARCH</li>
<li>
<code><a href="reference/filter.uc.html">filter.uc()</a></code> : Filter for cluster members</li>
<li>
<code><a href="reference/SimMatAbundance.html">SimMatAbundance()</a></code> : Compute histogram shape similarity between species</li>
</ul>
</div>
<div id="ltr-copy-number-estimation" class="section level4">
<h4 class="hasAnchor">
<a href="#ltr-copy-number-estimation" class="anchor"></a>LTR Copy Number Estimation</h4>
<ul>
<li>
<code><a href="reference/ltr.cn.html">ltr.cn()</a></code> : Detect solo LTR copies of predicted LTR transposons</li>
<li>
<code><a href="reference/cn2bed.html">cn2bed()</a></code> : Write copy number estimation results to BED file format.</li>
</ul>
</div>
<div id="filter-functions" class="section level4">
<h4 class="hasAnchor">
<a href="#filter-functions" class="anchor"></a>Filter Functions</h4>
<ul>
<li>
<code><a href="reference/filter.jumpers.html">filter.jumpers()</a></code> : Detect LTR retrotransposons that are potential jumpers</li>
<li>
<code><a href="reference/tidy.datasheet.html">tidy.datasheet()</a></code> : Select most important columns of ‘LTRpred’ output for further analytics</li>
</ul>
</div>
<div id="import-the-output-files-of-the-prediction-tools" class="section level4">
<h4 class="hasAnchor">
<a href="#import-the-output-files-of-the-prediction-tools" class="anchor"></a>Import the Output Files of the Prediction Tools:</h4>
<ul>
<li>
<code><a href="reference/read.prediction.html">read.prediction()</a></code> : Import the output of LTRharvest or LTRdigest</li>
<li>
<code><a href="reference/read.tabout.html">read.tabout()</a></code> : Import information sheet returned by LTRdigest</li>
<li>
<code><a href="reference/read.orfs.html">read.orfs()</a></code> : Read output of <code><a href="reference/ORFpred.html">ORFpred()</a></code>
</li>
<li>
<code><a href="reference/read.seqs.html">read.seqs()</a></code> : Import sequences of predicted LTR transposons</li>
<li>
<code><a href="reference/read.ltrpred.html">read.ltrpred()</a></code> : Import the data sheet file generated by <code><a href="reference/LTRpred.html">LTRpred()</a></code>
</li>
<li>
<code><a href="reference/read.uc.html">read.uc()</a></code> : Read file in USEARCH cluster format</li>
<li>
<code><a href="reference/read.blast6out.html">read.blast6out()</a></code> : Read file in blast6out format generated by USEARCH or VSEARCH</li>
</ul>
</div>
<div id="export-the-output-files-of-the-prediction-tools" class="section level4">
<h4 class="hasAnchor">
<a href="#export-the-output-files-of-the-prediction-tools" class="anchor"></a>Export the Output Files of the Prediction Tools:</h4>
<ul>
<li>
<code><a href="reference/pred2bed.html">pred2bed()</a></code> : Format LTR prediction data to BED file format</li>
<li>
<code><a href="reference/pred2fasta.html">pred2fasta()</a></code> : Save the sequence of the predicted LTR Transposons in a fasta file</li>
<li>
<code><a href="reference/pred2gff.html">pred2gff()</a></code> : Format LTR prediction data to GFF3 file format</li>
<li>
<code>pred2annotation()</code> : Match LTRharvest, LTRdigest, or LTRpred prediction with a given annotation file in GFF3 format</li>
<li>
<code><a href="reference/pred2csv.html">pred2csv()</a></code> : Format LTR prediction data to CSV file format</li>
</ul>
</div>
<div id="analytics-tools" class="section level4">
<h4 class="hasAnchor">
<a href="#analytics-tools" class="anchor"></a>Analytics Tools:</h4>
<ul>
<li>
<code><a href="reference/ORFpred.html">ORFpred()</a></code> : Open Reading Frame prediction in putative LTR transposons</li>
</ul>
</div>
<div id="annotation-and-validation" class="section level4">
<h4 class="hasAnchor">
<a href="#annotation-and-validation" class="anchor"></a>Annotation and Validation:</h4>
<ul>
<li>
<code><a href="reference/dfam.query.html">dfam.query()</a></code> : Annotation of <code>de novo</code> predicted LTR transposons via <a href="http://dfam.org/help/tools">Dfam</a> searches</li>
<li>
<code><a href="reference/read.dfam.html">read.dfam()</a></code> : Import Dfam Query Output</li>
<li>
<code><a href="reference/repbase.clean.html">repbase.clean()</a></code> : Clean the initial Repbase database for BLAST</li>
<li>
<code><a href="reference/repbase.query.html">repbase.query()</a></code> : Query the RepBase to annotate putative LTRs</li>
<li>
<code><a href="reference/repbase.filter.html">repbase.filter()</a></code> : Filter the Repbase query output</li>
</ul>
</div>
<div id="methylation-context-estimation" class="section level4">
<h4 class="hasAnchor">
<a href="#methylation-context-estimation" class="anchor"></a>Methylation Context Estimation</h4>
<ul>
<li>
<code><a href="reference/motif.count.html">motif.count()</a></code> : Low level function to detect motifs in strings</li>
</ul>
</div>
<div id="visualization-framework" class="section level4">
<h4 class="hasAnchor">
<a href="#visualization-framework" class="anchor"></a>Visualization Framework</h4>
<ul>
<li>
<code>plot_ltrsim_individual()</code> : Plot the age distribution of predicted LTR transposons</li>
<li>
<code><a href="reference/plot_ltrwidth_individual.html">plot_ltrwidth_individual()</a></code> : Plot the width distribution of putative LTR transposons or LTRs for individual species</li>
<li>
<code><a href="reference/plot_ltrwidth_species.html">plot_ltrwidth_species()</a></code> : Plot the width distribution of putative LTR transposons or LTRs for all species</li>
<li>
<code><a href="reference/plot_ltrwidth_kingdom.html">plot_ltrwidth_kingdom()</a></code> : Plot the width distribution of putative LTR transposons or LTRs for all kingdoms</li>
<li>
<code><a href="reference/plot_copynumber_individual.html">plot_copynumber_individual()</a></code> : Plot the copy number distribution of putative LTR transposons or LTRs for individual species</li>
<li>
<code><a href="reference/plot_copynumber_species.html">plot_copynumber_species()</a></code> : Plot the copy number distribution of putative LTR transposons or LTRs for all species</li>
<li>
<code><a href="reference/plot_copynumber_kingdom.html">plot_copynumber_kingdom()</a></code> : Plot the copy number distribution of putative LTR transposons or LTRs for all kingdoms</li>
<li>
<code><a href="reference/plotLTRRange.html">plotLTRRange()</a></code> : Plot Genomic Ranges of putative LTR transposons</li>
<li>
<code>PlotSimCount()</code> : Plot LTR Similarity vs. predicted LTR count</li>
<li>
<code><a href="reference/plotSize.html">plotSize()</a></code> : Plot Genome size vs. LTR transposon count</li>
<li>
<code><a href="reference/plotSizeJumpers.html">plotSizeJumpers()</a></code> : Plot Genome size vs. LTR transposon count for jumpers</li>
<li>
<code><a href="reference/plotFamily.html">plotFamily()</a></code> : Visualize the Superfamily distribution of predicted LTR retrotransposons</li>
<li>
<code><a href="reference/plotDomain.html">plotDomain()</a></code> : Visualize the Protein Domain distribution of predicted LTR retrotransposons</li>
<li>
<code><a href="reference/plotCN.html">plotCN()</a></code> : Plot correlation between LTR copy number and methylation context<br>
</li>
<li>
<code><a href="reference/plotCluster.html">plotCluster()</a></code> : Plot correlation between Cluster Number and any other variable</li>
<li>
<code>PlotInterSpeciesCluster()</code> : Plot inter species similarity between TEs (for a specific cluster)</li>
<li>
<code>PlotMainInterSpeciesCluster()</code> : Plot inter species similarity between TEs (for the top n clusters)</li>
</ul>
</div>
<div id="minor-helper-functions" class="section level4">
<h4 class="hasAnchor">
<a href="#minor-helper-functions" class="anchor"></a>Minor helper functions</h4>
<ul>
<li>
<code><a href="reference/bcolor.html">bcolor()</a></code> : Beautiful colors for plots</li>
<li>
<code><a href="reference/file.move.html">file.move()</a></code> : Move folders from one location to another</li>
<li>
<code><a href="reference/get.pred.filenames.html">get.pred.filenames()</a></code> : Retrieve file names of files genereated by LTRpred</li>
<li>
<code><a href="reference/get.seqs.html">get.seqs()</a></code> : Quickly retrieve the sequences of a ‘Biostrings’ object</li>
<li>
<code><a href="reference/ws.wrap.path.html">ws.wrap.path()</a></code> : Wrap whitespace in paths</li>
<li>
<code><a href="reference/rename.fasta.html">rename.fasta()</a></code> : rename.fasta</li>
</ul>
</div>
</div>
</div>
<div id="acknowledgement" class="section level2">
<h2 class="hasAnchor">
<a href="#acknowledgement" class="anchor"></a>Acknowledgement</h2>
<p>I would like to thank the <a href="http://www.slcu.cam.ac.uk/research/paszkowski-group/group-members">Paszkowski team</a> for incredible support and motivating discussions that led to the realization of this project.</p>
</div>

  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <div class="links">
<h2>Links</h2>
<ul class="list-unstyled">
<li>Browse source code at <br><a href="https://github.com/HajkD/LTRpred">https://​github.com/​HajkD/​LTRpred</a>
</li>
<li>Report a bug at <br><a href="https://github.com/HajkD/LTRpred/issues">https://​github.com/​HajkD/​LTRpred/​issues</a>
</li>
</ul>
</div>
<div class="license">
<h2>License</h2>
<ul class="list-unstyled">
<li><a href="https://www.r-project.org/Licenses/GPL-2">GPL-2</a></li>
</ul>
</div>
<div class="developers">
<h2>Developers</h2>
<ul class="list-unstyled">
<li>Hajk-Georg Drost <br><small class="roles"> Author, maintainer </small>  </li>
</ul>
</div>

  </div>
</div>


      <footer><div class="copyright">
  <p>Developed by Hajk-Georg Drost.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
