<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Summary • LTRpred</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script>




<meta property="og:title" content="Summary" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-title-body">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">LTRpred</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="articles/Introduction.html">Introduction to LTRpred</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/HajkD/LTRpred/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="contents col-md-9">
    <div class="page-header">
      <h1>Summary</h1>
    </div>


<hr>
<p>title: ‘LTRpred: <em>de novo</em> annotation of intact retrotransposons’ authors:</p>
<ul>
<li>affiliation: “1, 2” name: Hajk-Georg Drost orcid: 0000-0002-1567-306X date: “28 February 2020” output: pdf_document bibliography: paper.bib tags:</li>
<li>R</li>
<li>retrotransposon annotation</li>
<li>de novo functional annotation of retrotransposons</li>
<li>meta-genome scale retrotransposon annotation</li>
<li>transposable elements</li>
<li>mobile transposable elements</li>
<li>genome dynamics affiliations:</li>
<li>index: 1 name: The Sainsbury Laboratory, University of Cambridge, Bateman Street, Cambridge CB2 1LR, UK</li>
<li>index: 2 name: Department of Molecular Biology, Max Planck Institute for Developmental Biology, 72076 Tuebingen, Germany —</li>
</ul>
<div id="summary" class="section level1">

<p>Transposable elements (TEs) play a crucial role in altering the genomic landscape of all organisms and thereby massively influence the genetic information passed on to succeeding generations [@Sundaram2020-yw]. In the past, TEs were seen as selfish mobile elements populating host genomes to increase their chances for transgenerational transmission over long evolutionary time scales. This notion of selfish elements is slowly changing [@Drost2019-rz] and a new picture drawing a complex genetic landscape benefitting both, host and TE, emerges whereby novel forms can arise through random shuffling of genetic material. For example, the tomato fruit shape [@Benoit2019-ux], moth adaptive cryptic coloration that occurred during the industrial revolution [@Chuong2017-cb], and inner cell mass development in human embryonic stem cells [@Chuong2017-cb] were all shown to be driven by TE activity. Thus, the impact of these elements on altering morphological traits is imminent and requires new attention in the light of evolvability. However, TEs tend to degenerate their sequence leaving their fragmented copies considered as <em>junk</em> DNA in host genomes, which hamper assembly and annotation of new genomes.</p>
<p>Nowadays, the <em>de novo</em> detection of transposable elements is performed by annotation tools specifically designed to capture any type of repeated sequence, TE family, or remnant DNA loci that can be associated with known transposable elements within a genome assembly. The main goal of such efforts is to retrieve a maximum number of loci that can be associated with known TEs. If successful, such annotation can then be used to mask host genomes from TE remnants to simplify genomics studies focusing on host genes. Therefore, there is no automatically performed distinction between complete and potentially active TE and their mutated copies.</p>
<p>Here, we introduce the <code>LTRpred</code> pipeline which allows to <em>de novo</em> annotate functional and thus potentially mobile retrotransposons in any given genome assembly. Different from other annotation tools, <code>LTRpred</code> focuses on retrieving structurally intact elements within sequences of genomes rather than characterizing all traces of historic TE activity.</p>
<p>Such functional annotation is most useful when trying to spot retrotransposons responsible for recent reshuffling of genetic material in the tree of life. Detecting and further characterization of those active retrotransposons yields the potential to harness them as mutagenesis agents by inducing transposition bursts in a controlled fashion to stimulate genomic reshaping processes towards novel traits.</p>
</div>
<div id="ltrpred-emerged-as-valuable-tool-for-diverse-te-mobilization-studies" class="section level1">
<h1 class="hasAnchor">
<a href="#ltrpred-emerged-as-valuable-tool-for-diverse-te-mobilization-studies" class="anchor"></a><code>LTRpred</code> emerged as valuable tool for diverse TE mobilization studies</h1>
<p><code>LTRpred</code> was successfully used in previous studies to annotate functional retrotransposons for various applications. In detail, <code>LTRpred</code> was used to annotate the retrotransposon family <code>RIDER</code> within the plant kingdom, shown to be involved in tomato fruit shape elongation. Together with experimental evidence, our analyses revealed that RIDER elements can be activated via drought stress and may help plants rich in RIDER activity to better adapt to drought stress conditions [@Benoit2019-ux]. In a complementary study, <code>LTRpred</code> was used to generate a candidate list of potentially mobile retrotransposon families in rice and tomato, which were then confirmed to produce extrachromosomal DNA using the ALE-Seq methodology [@Cho2019-zp]. Finally, <code>LTRpred</code> supported efforts to annotate and date functional retrotransposons in tomato and <em>Arabidopsis</em> which led to the finding that chromodomain DNA methyltransferases (CMTs) silence young and intact retrotransposons in distal chromatin whereas older non-functional retrotransposons are affected by small RNA-directed DNA methylation [@Wang2019].</p>
<p>Together, potentially functional retrotransposons annotated <em>de novo</em> with <code>LTRpred</code> were subsequently shown to be active and mobile in diverse molecular studies. This approach may stimulate a new wave of research towards understanding the physiological role of functional retrotransposons and to reveal the mechanistic principles of transposon associated evolvability.</p>
</div>
<div id="pipeline-dependencies" class="section level1">
<h1 class="hasAnchor">
<a href="#pipeline-dependencies" class="anchor"></a>Pipeline dependencies</h1>
<p>The LTRpred pipeline depends on the R packages biomartr [@Drost2017-cw], dplyr [@Wickham2019-kh], Biostrings, stringr [@Wickham2019-kh], IRanges [@Lawrence2013-dv], RColorBrewer, ggplot2 [@Wickham2016-eq], readr [@Wickham2019-kh], magrittr, downloader, BSDA, ggrepel, ggbio [@Yin2012-ro], and gridExtra.</p>
<p>In detail, the LTRpred pipeline calls the command line tools suffixerator, LTRharvest [@Ellinghaus2008-hu], and LTRdigest [@Steinbiss2009-vg], which are part of the GenomeTools library [@Gremme2013-ba] using customized parameter settings to screen for repeated LTRs, specific sequence motifs such as primer binding sites (PBS), polypurine tract motifs (PPT), and target site duplications (TSD) and for conserved protein domains such as reverse transcriptase (gag), integrase DNA binding domain, integrase Zinc binding domain, RNase H, and the integrase core domain. The LTRharvest and LTRdigest outputs are efficiently parsed by LTRpred and transformed into a tidy data format [@Wickham2019-kh] which subsequently enables automation of false positive curation. Next, open reading frame (ORF) prediction is performed by a customized wrapper function that runs the command line tool usearch [@Edgar2010-cb]. This step allows to automatically filter out retrotransposons that might have conserved protein domains such as an integrase or reverse transcriptase, but fail to have any ORFs and thus might not be expressed. In a third step, retrotransposon family clustering is performed using sequence clustering with vsearch [@Rognes2016-sk] which defines family members by &gt;90% sequence homology of the full element to each other. In a fourth step, an automated hmmer search [@Finn2011-fs] against the Dfam database (Hubley et al., 2016) is performed to assign super-family associations such as Copia or Gypsy by comparing the protein domains of de novo predicted retrotransposons with already annotated TEs in the Dfam (<a href="https://dfam.org/home" class="uri">https://dfam.org/home</a>) database. In the last step, the <em>de novo</em> annotated 5 prime and 3 prime LTR sequences are used to estimate the evolutionary age of the retrotransposon which should be treated with caution since retrotransposons can undergo reverse-transcriptase mediated recombination [@Sanchez2017-sy].</p>
</div>
<div id="example-workflow" class="section level1">
<h1 class="hasAnchor">
<a href="#example-workflow" class="anchor"></a>Example workflow</h1>
<p>After installing all prerequisite command line tools (<a href="https://hajkd.github.io/LTRpred/articles/Introduction.html#installation" class="uri">https://hajkd.github.io/LTRpred/articles/Introduction.html#installation</a>) users can run the <code><a href="reference/LTRpred.html">LTRpred()</a></code> pipeline using the default parameter configuration. In the following example, an LTR transposon prediction is performed for parts of the Human Y chromosome.</p>
<div class="sourceCode" id="cb1"><pre class="downlit">
<span class="co"># load LTRpred package</span>
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="kw"><a href="https://github.com/HajkD/LTRpred">LTRpred</a></span>)
<span class="co"># de novo LTR transposon prediction for the Human Y chromosome</span>
<span class="fu"><a href="reference/LTRpred.html">LTRpred</a></span>(
    genome.file = <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="st">"Hsapiens_ChrY.fa"</span>, package = <span class="st">"LTRpred"</span>),
    cores = <span class="fl">4</span>
)
</pre></div>
<p>The LTRpred() output table <code>*_LTRpred_DataSheet.tsv</code> is in tidy format and can then be imported using <code><a href="reference/read.ltrpred.html">read.ltrpred()</a></code>. The tidy output format is designed to work seamlessly with the tidyverse and R data science framework.</p>
<div class="sourceCode" id="cb2"><pre class="downlit">
<span class="co"># import LTRpred prediction output</span>
<span class="kw">Hsapiens_chrY</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/read.ltrpred.html">read.ltrpred</a></span>(<span class="st">"Hsapiens_ChrY_ltrpred/
Hsapiens_ChrY_LTRpred_DataSheet.tsv"</span>)
<span class="co"># look at some results</span>
<span class="kw">dplyr</span>::<span class="fu"><a href="https://tibble.tidyverse.org/reference/glimpse.html">glimpse</a></span>(<span class="kw">Hsapiens_chrY</span>)
</pre></div>
<pre><code>Observations: 21
Variables: 92
$ species                 &lt;chr&gt; "Hsapiens_ChrY", "Hsapien...
$ ID                      &lt;chr&gt; "Hsapiens_ChrY_LTR_retrot...
$ dfam_target_name        &lt;chr&gt; NA, NA, NA, NA, NA, NA, N...
$ ltr_similarity          &lt;dbl&gt; 80.73, 89.85, 79.71, 83.2...
$ ltr_age_mya             &lt;dbl&gt; 0.7936246, 0.2831139, 0.7...
$ similarity              &lt;chr&gt; "(80,82]", "(88,90]", "(7...
$ protein_domain          &lt;chr&gt; "RVT_1", "RVT_1", NA, NA,...
$ orfs                    &lt;int&gt; 1, 1, 0, 0, 0, 0, 0, 1, 0...
$ chromosome              &lt;chr&gt; "NC000024.10Homosa", "NC0...
$ start                   &lt;int&gt; 3143582, 3275798, 3313536...
$ end                     &lt;int&gt; 3162877, 3299928, 3318551...
$ strand                  &lt;chr&gt; "-", "-", "+", "+", "-", ...
$ width                   &lt;int&gt; 19296, 24131, 5016, 12952...
$ annotation              &lt;chr&gt; "LTR_retrotransposon", "L...
$ pred_tool               &lt;chr&gt; "LTRpred", "LTRpred", "LT...
$ frame                   &lt;chr&gt; ".", ".", ".", ".", ".", ...
$ score                   &lt;chr&gt; ".", ".", ".", ".", ".", ...
$ lLTR_start              &lt;int&gt; 3143582, 3275798, 3313536...
$ lLTR_end                &lt;int&gt; 3143687, 3276408, 3313665...
$ lLTR_length             &lt;int&gt; 106, 611, 130, 126, 218, ...
$ rLTR_start              &lt;int&gt; 3162769, 3299338, 3318414...
$ rLTR_end                &lt;int&gt; 3162877, 3299928, 3318551...
$ rLTR_length             &lt;int&gt; 109, 591, 138, 137, 219, ...
$ lTSD_start              &lt;int&gt; 3143578, 3275794, 3313532...
$ lTSD_end                &lt;int&gt; 3143581, 3275797, 3313535...
$ lTSD_motif              &lt;chr&gt; "acag", "ttgt", "ttag", "...
$ rTSD_start              &lt;int&gt; 3162878, 3299929, 3318552...
$ rTSD_end                &lt;int&gt; 3162881, 3299932, 3318555...
$ rTSD_motif              &lt;chr&gt; "acag", "ttgt", "ttag", "...
$ PPT_start               &lt;int&gt; NA, NA, NA, NA, NA, 34660...
$ PPT_end                 &lt;int&gt; NA, NA, NA, NA, NA, 34660...
$ PPT_motif               &lt;chr&gt; NA, NA, NA, NA, NA, "agag...
$ PPT_strand              &lt;chr&gt; NA, NA, NA, NA, NA, "+", ...
$ PPT_offset              &lt;int&gt; NA, NA, NA, NA, NA, 23, N...
$ PBS_start               &lt;int&gt; NA, NA, 3313667, 3372512,...
$ PBS_end                 &lt;int&gt; NA, NA, 3313677, 3372522,...
$ PBS_strand              &lt;chr&gt; NA, NA, "+", "+", "-", "+...
$ tRNA                    &lt;chr&gt; NA, NA, "Homo_sapiens_tRN...
$ tRNA_motif              &lt;chr&gt; NA, NA, "aattagctgga", "c...
$ PBS_offset              &lt;int&gt; NA, NA, 1, 3, 0, 5, 2, 5,...
$ tRNA_offset             &lt;int&gt; NA, NA, 1, 0, 2, 5, 1, 5,...
$ `PBS/tRNA_edist`        &lt;int&gt; NA, NA, 1, 1, 1, 1, 1, 1,...
$ orf.id                  &lt;chr&gt; "NC000024.10Homosa_314358...
$ repeat_region_length    &lt;int&gt; 19304, 24139, 5024, 12960...
$ PPT_length              &lt;int&gt; NA, NA, NA, NA, NA, 27, N...
$ PBS_length              &lt;int&gt; NA, NA, 11, 11, 11, 11, 1...
$ dfam_acc                &lt;chr&gt; NA, NA, NA, NA, NA, NA, N...
$ dfam_bits               &lt;dbl&gt; NA, NA, NA, NA, NA, NA, N...
$ dfam_e_value            &lt;dbl&gt; NA, NA, NA, NA, NA, NA, N...
$ dfam_bias               &lt;dbl&gt; NA, NA, NA, NA, NA, NA, N...
$ `dfam_hmm-st`           &lt;dbl&gt; NA, NA, NA, NA, NA, NA, N...
$ `dfam_hmm-en`           &lt;dbl&gt; NA, NA, NA, NA, NA, NA, N...
$ dfam_strand             &lt;chr&gt; NA, NA, NA, NA, NA, NA, N...
$ `dfam_ali-st`           &lt;dbl&gt; NA, NA, NA, NA, NA, NA, N...
$ `dfam_ali-en`           &lt;dbl&gt; NA, NA, NA, NA, NA, NA, N...
$ `dfam_env-st`           &lt;dbl&gt; NA, NA, NA, NA, NA, NA, N...
$ `dfam_env-en`           &lt;dbl&gt; NA, NA, NA, NA, NA, NA, N...
$ dfam_modlen             &lt;dbl&gt; NA, NA, NA, NA, NA, NA, N...
$ dfam_target_description &lt;chr&gt; NA, NA, NA, NA, NA, NA, N...
$ Clust_Cluster           &lt;chr&gt; NA, NA, NA, NA, NA, NA, N...
$ Clust_Target            &lt;chr&gt; NA, NA, NA, NA, NA, NA, N...
$ Clust_Perc_Ident        &lt;dbl&gt; NA, NA, NA, NA, NA, NA, N...
$ Clust_cn                &lt;int&gt; NA, NA, NA, NA, NA, NA, N...
$ TE_CG_abs               &lt;dbl&gt; 62, 125, 35, 70, 139, 83,...
$ TE_CG_rel               &lt;dbl&gt; 0.003213101, 0.005180059,...
$ TE_CHG_abs              &lt;dbl&gt; 659, 830, 150, 396, 742, ...
$ TE_CHG_rel              &lt;dbl&gt; 0.03415216, 0.03439559, 0...
$ TE_CHH_abs              &lt;dbl&gt; 2571, 3454, 748, 1743, 31...
$ TE_CHH_rel              &lt;dbl&gt; 0.1332400, 0.1431354, 0.1...
$ TE_CCG_abs              &lt;dbl&gt; 13, 24, 6, 15, 33, 16, 4,...
$ TE_CCG_rel              &lt;dbl&gt; 0.0006737148, 0.000994571...
$ TE_N_abs                &lt;dbl&gt; 0, 0, 0, 0, 0, 0, 0, 0, 0...
$ CG_3ltr_abs             &lt;dbl&gt; 1, 0, 1, 4, 8, 3, 1, 11, ...
$ CG_3ltr_rel             &lt;dbl&gt; 0.009433962, 0.000000000,...
$ CHG_3ltr_abs            &lt;dbl&gt; 2, 24, 9, 8, 14, 14, 9, 9...
$ CHG_3ltr_rel            &lt;dbl&gt; 0.01886792, 0.03927987, 0...
$ CHH_3ltr_abs            &lt;dbl&gt; 18, 69, 18, 26, 43, 70, 2...
$ CHH_3ltr_rel            &lt;dbl&gt; 0.16981132, 0.11292962, 0...
$ CCG_3ltr_abs            &lt;dbl&gt; 0, 0, 0, 2, 2, 0, 1, 4, 5...
$ CCG_3ltr_rel            &lt;dbl&gt; 0.000000000, 0.000000000,...
$ N_3ltr_abs              &lt;dbl&gt; 0, 0, 0, 0, 0, 0, 0, 0, 0...
$ CG_5ltr_abs             &lt;dbl&gt; 1, 0, 1, 4, 8, 3, 1, 11, ...
$ CG_5ltr_rel             &lt;dbl&gt; 0.009433962, 0.000000000,...
$ CHG_5ltr_abs            &lt;dbl&gt; 2, 24, 9, 8, 14, 14, 9, 9...
$ CHG_5ltr_rel            &lt;dbl&gt; 0.01886792, 0.03927987, 0...
$ CHH_5ltr_abs            &lt;dbl&gt; 18, 69, 18, 26, 43, 70, 2...
$ CHH_5ltr_rel            &lt;dbl&gt; 0.16981132, 0.11292962, 0...
$ CCG_5ltr_abs            &lt;dbl&gt; 0, 0, 0, 2, 2, 0, 1, 4, 5...
$ CCG_5ltr_rel            &lt;dbl&gt; 0.000000000, 0.000000000,...
$ N_5ltr_abs              &lt;dbl&gt; 0, 0, 0, 0, 0, 0, 0, 0, 0...
$ cn_3ltr                 &lt;dbl&gt; NA, NA, NA, NA, NA, NA, N...
$ cn_5ltr                 &lt;dbl&gt; NA, NA, NA, NA, NA, NA, N..</code></pre>
<div id="ltrpred-output" class="section level3">
<h3 class="hasAnchor">
<a href="#ltrpred-output" class="anchor"></a>LTRpred output</h3>
<p>The <code><a href="reference/LTRpred.html">LTRpred()</a></code> function internally generates a folder named <code>*_ltrpred</code> which stores all output annotation and sequence files.</p>
<p>In detail, the following files and folders are generated by the <code><a href="reference/LTRpred.html">LTRpred()</a></code> function:</p>
<ul>
<li>
<strong>Folder <code>*_ltrpred</code></strong>
<ul>
<li><p><code>*_ORF_prediction_nt.fsa</code> : Stores the predicted open reading frames within the predicted LTR transposons as DNA sequence.</p></li>
<li><p><code>*_ORF_prediction_aa.fsa</code> : Stores the predicted open reading frames within the predicted LTR transposons as protein sequence.</p></li>
<li><p><code>*_LTRpred.gff</code> : Stores the LTRpred predicted LTR transposons in GFF format.</p></li>
<li><p><code>*_LTRpred.bed</code> : Stores the LTRpred predicted LTR transposons in BED format.</p></li>
<li><p><code>*_LTRpred_DataSheet.tsv</code> : Stores the output table as data sheet.</p></li>
<li>
<p><strong>Folder <code>*_ltrharvest</code></strong></p>
<ul>
<li><p><code>*_ltrharvest/*_BetweenLTRSeqs.fsa</code> : DNA sequences of the region between the LTRs in fasta format.</p></li>
<li><p><code>*_ltrharvest/*_Details.tsv</code> : A spread sheet containing detailed information about the predicted LTRs.</p></li>
<li><p><code>*_ltrharvest/*_FullLTRRetrotransposonSeqs.fsa</code> : DNA sequences of the entire predicted LTR retrotransposon.</p></li>
<li><p><code>*_ltrharvest/*_index.fsa</code> : The suffixarray index file used to predict putative LTR retrotransposons.</p></li>
<li><p><code>*_ltrharvest/*_Prediction.gff</code> : A spread sheet containing detailed additional information about the predicted LTRs (partially redundant with the *_Details.tsv file).</p></li>
</ul>
</li>
<li>
<p><strong>Folder <code>*_ltrdigest</code></strong></p>
<ul>
<li><p><code>*_ltrdigest/*_LTRdigestPrediction.gff</code> : A spread sheet containing detailed information about the predicted LTRs.</p></li>
<li><p><code>*_ltrdigest/*_index_ltrdigest.fsa</code> : The suffixarray index file used to predict putative LTR retrotransposons with LTRdigest.</p></li>
<li><p><code>*_ltrdigest/*-ltrdigest_tabout.csv</code> : A spread sheet containing additional detailed information about the predicted LTRs.</p></li>
<li><p><code>*_ltrdigest/*-ltrdigest_complete.fas</code> : The full length DNA sequences of all predicted LTR transposons.</p></li>
<li><p><code>*_ltrdigest/*-ltrdigest_conditions.csv</code> : Contains information about the parameters used for a given LTRdigest run.</p></li>
<li><p><code>*_ltrdigest/*-ltrdigest_pbs.fas</code> : Stores the predicted PBS sequences for the putative LTR retrotransposons.</p></li>
<li><p><code>*_ltrdigest/*-ltrdigest_ppt.fas</code> : Stores the predicted PPT sequences for the putative LTR retrotransposons.</p></li>
<li><p><code>*_ltrdigest/*-ltrdigest_5ltr.fas</code> and <code>*-ltrdigest_3ltr.fas</code>: Stores the predicted 5’ and 3’ LTR sequences. Note: If the direction of the putative retrotransposon could be predicted, these files will contain the corresponding 3’ and 5’ LTR sequences. If no direction could be predicted, forward direction with regard to the original sequence will be assumed by LTRdigest, i.e. the ‘left’ LTR will be considered the 5’ LTR.</p></li>
<li><p><code>*_ltrdigest/*-ltrdigest_pdom_&lt;domainname&gt;.fas</code> : Stores the DNA sequences of the HMM matches to the LTR retrotransposon candidates.</p></li>
<li><p><code>*_ltrdigest/*-ltrdigest_pdom_&lt;domainname&gt;_aa.fas</code> : Stores the concatenated protein sequences of the HMM matches to the LTR retrotransposon candidates.</p></li>
<li><p><code>*_ltrdigest/*-ltrdigest_pdom_&lt;domainname&gt;_ali.fas</code> : Stores the alignment information for all matches of the given protein domain model to the translations of all candidates.</p></li>
</ul>
</li>
</ul>
</li>
</ul>
</div>
<div id="visualising-functional-retrotransposons-annotated-with-ltrpred" class="section level3">
<h3 class="hasAnchor">
<a href="#visualising-functional-retrotransposons-annotated-with-ltrpred" class="anchor"></a>Visualising functional retrotransposons annotated with <code>LTRpred</code>
</h3>
<p>Finally, users can visualise the positioning of <em>de novo</em> annotated retrotransposons along the chromosomes. Here, we choose an example based on the yeast genome.</p>
<div class="sourceCode" id="cb4"><pre class="downlit">
<span class="co"># install.packages("biomartr")</span>
<span class="co"># download the yeast genome from ENSEMBL</span>
<span class="kw">sc_genome_path</span> <span class="op">&lt;-</span> <span class="kw">biomartr</span>::<span class="fu"><a href="https://docs.ropensci.org/biomartr/reference/getGenome.html">getGenome</a></span>(db = <span class="st">"ensembl"</span>, 
                               organism = <span class="st">"Saccharomyces cerevisiae"</span>, 
                               path = <span class="st">"yeast_genome"</span>, 
                               gunzip = <span class="fl">TRUE</span>)
<span class="co"># run LTRpred on the yeast genome (-&gt; this may take a few minutes)</span>
<span class="kw">LTRpred</span>::<span class="fu"><a href="reference/LTRpred.html">LTRpred</a></span>(
    genome.file = <span class="kw">sc_genome_path</span>,
    cores = <span class="fl">4</span>
)
<span class="co"># import functional retrotransposon annotation</span>
<span class="kw">sc_ltrpred</span> <span class="op">&lt;-</span> <span class="kw">LTRpred</span>::<span class="fu"><a href="reference/read.ltrpred.html">read.ltrpred</a></span>(
 <span class="fu"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span>(<span class="st">"Saccharomyces_cerevisiae_ltrpred"</span>,
                      <span class="st">"Saccharomyces_cerevisiae_LTRpred_DataSheet.tsv"</span>))
<span class="co"># filter for potentially active candidates</span>
<span class="kw">sc_ltrpred_filtered</span> <span class="op">&lt;-</span> <span class="kw">LTRpred</span>::<span class="fu"><a href="reference/quality.filter.html">quality.filter</a></span>(<span class="kw">sc_ltrpred</span>, 
                                          sim = <span class="fl">95</span>, 
                                          strategy = <span class="st">"stringent"</span>, 
                                          n.orfs = <span class="fl">1</span>)
<span class="co"># visualise the position of potentially active retrotransposons</span>
<span class="co"># along the yeast chromosomes</span>
<span class="kw">LTRpred</span>::<span class="fu"><a href="reference/plot_element_distr_along_chromosome.html">plot_element_distr_along_chromosome</a></span>(<span class="kw">sc_ltrpred_filtered</span>,
                                             <span class="kw">sc_genome_path</span>)
</pre></div>
<p><img src="Sc_LTRpred.png" alt="Positions of functional retrotransposons annotated by LTRpred along the yeast chromosomes.">{ width=70% }</p>
</div>
<div id="metagenome-scale-annotations" class="section level3">
<h3 class="hasAnchor">
<a href="#metagenome-scale-annotations" class="anchor"></a>Metagenome scale annotations</h3>
<p><code>LTRpred</code> allows users to generate annotations not only for single genomes but for multiple genomes (metagenomes) using only one pipeline function named <code><a href="reference/LTRpred.meta.html">LTRpred.meta()</a></code>.</p>
<p>Users can download the <code>biomartr</code> package [@Drost2017-cw] to automatically retrieve genome assembly files for the species of interest.</p>
<div class="sourceCode" id="cb5"><pre class="downlit">
<span class="co"># specify the scientific names of the species of interest</span>
<span class="co"># that shall first be downloaded and then be used</span>
<span class="co"># to generate LTRpred annotations</span>
<span class="kw">species</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"Arabidopsis thaliana"</span>, <span class="st">"Arabidopsis lyrata"</span>, <span class="st">"Capsella rubella"</span>) 
<span class="co"># download the genome assembly files for the species of interest</span>
<span class="co"># </span>
<span class="co"># install.packages("biomartr")</span>
<span class="kw">biomartr</span>::<span class="fu"><a href="https://docs.ropensci.org/biomartr/reference/getGenomeSet.html">getGenomeSet</a></span>(db = <span class="st">"refseq"</span>, organisms = <span class="kw">species</span>,
                      path = <span class="st">"store_genome_set"</span>)
<span class="co"># run LTRpred.meta() on the 3 species with 3 cores</span>
<span class="kw">LTRpred</span>::<span class="fu"><a href="reference/LTRpred.meta.html">LTRpred.meta</a></span>(genome.folder = <span class="st">"store_genome_set"</span>,
                      output.folder = <span class="st">"LTRpred_meta_results"</span>,
                      cores = <span class="fl">3</span>)
</pre></div>
</div>
</div>
<div id="acknowledgements" class="section level1">
<h1 class="hasAnchor">
<a href="#acknowledgements" class="anchor"></a>Acknowledgements</h1>
<p>This work was supported by a European Research Council grant named EVOBREED [grant number 322621] and a Gatsby Fellowship [grant number AT3273/GLE]. The author also thanks Jerzy Paszkowski for supporting the development and application of <code>LTRpred</code> and for providing valuable suggestions to improve the manuscript.</p>
</div>
<div id="references" class="section level1">
<h1 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h1>
</div>


  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>

</div>



      <footer>
      <div class="copyright">
  <p>Developed by Hajk-Georg Drost.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


