<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html>
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Function reference â€¢ LTRpred</title>

<!-- jquery -->
<script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/1.7.1/clipboard.min.js" integrity="sha384-cV+rhyOuRHc9Ub/91rihWcGmMmCXDeksTtCihMupQHSsi8GIIRDG0ThDc3HGQFJ3" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script>
<script src="../pkgdown.js"></script>
<meta property="og:title" content="Function reference" />


<!-- mathjax -->
<script src='https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->




  </head>

  <body>
    <div class="container template-reference-index">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">LTRpred</a>
        <span class="label label-default" data-toggle="tooltip" data-placement="bottom" title="Released package">1.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/Introduction.html">Introduction to LTRpred</a>
    </li>
  </ul>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/HajkD/LTRpred">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

<div class="row">
  <div class="contents col-md-9">
    <div class="page-header">
      <h1>Reference</h1>
    </div>

    <table class="ref-index">

    <colgroup>
      <col class="alias" />
      <col class="title" />
    </colgroup>

    <tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-all-functions" class="hasAnchor"><a href="#section-all-functions" class="anchor"></a>All functions</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      <tr>
        <!--  -->
        <td>
          <p><code><a href="AllPairwiseAlign.html">AllPairwiseAlign()</a></code> </p>
        </td>
        <td><p>Compute all pairwise (global) alignments with VSEARCH</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="CLUSTpred.html">CLUSTpred()</a></code> </p>
        </td>
        <td><p>Cluster Sequences with VSEARCH</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="LTRdigest.html">LTRdigest()</a></code> </p>
        </td>
        <td><p>Run LTRdigest to predict putative LTR Retrotransposons</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="LTRharvest.html">LTRharvest()</a></code> </p>
        </td>
        <td><p>Run LTRharvest to predict putative LTR Retrotransposons</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="LTRpred.html">LTRpred()</a></code> </p>
        </td>
        <td><p>Predict LTR retrotransposons in a given genome</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="LTRpred.meta.html">LTRpred.meta()</a></code> </p>
        </td>
        <td><p>Perform Meta-Analyses with LTRpred</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="ORFpred.html">ORFpred()</a></code> </p>
        </td>
        <td><p>Open Reading Frame Prediction</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="SimMatAbundance.html">SimMatAbundance()</a></code> </p>
        </td>
        <td><p>Compute histogram shape similarity between species</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="assign.short.sci.name.html">assign.short.sci.name()</a></code> </p>
        </td>
        <td><p>Transform long file names to short scientific names</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="bcolor.html">bcolor()</a></code> </p>
        </td>
        <td><p>Colors scheme for plots</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="clust2fasta.html">clust2fasta()</a></code> </p>
        </td>
        <td><p>Export sequences of TEs belonging to the same cluster</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="cluster.members.html">cluster.members()</a></code> </p>
        </td>
        <td><p>Select members of a specific cluster</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="cn2bed.html">cn2bed()</a></code> </p>
        </td>
        <td><p>Write copy number estimation results to BED file format.</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="combinePreds.html">combinePreds()</a></code> </p>
        </td>
        <td><p>Combine LTRpred prediction files</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="dfam.query.html">dfam.query()</a></code> </p>
        </td>
        <td><p>Query the Dfam database to annotate putative LTRs</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="file.move.html">file.move()</a></code> </p>
        </td>
        <td><p>Move folders from one location to another</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="filter.jumpers.html">filter.jumpers()</a></code> </p>
        </td>
        <td><p>Detect LTR retrotransposons that are potential jumpers</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="filter.uc.html">filter.uc()</a></code> </p>
        </td>
        <td><p>Filter for cluster members</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="generate.multi.quality.filter.meta.html">generate.multi.quality.filter.meta()</a></code> </p>
        </td>
        <td><p>Run <code>quality.filter.meta</code> for several different ltr similarity thresholds.</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="genome.summary.html">genome.summary()</a></code> </p>
        </td>
        <td><p>Generating genome summary files for <code>LTRpred.meta</code> results</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="get.pred.filenames.html">get.pred.filenames()</a></code> </p>
        </td>
        <td><p>Retrieve file names of files genereated by LTRpred</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="get.seqs.html">get.seqs()</a></code> </p>
        </td>
        <td><p>Quickly retrieve the sequences of a <code>Biostrings</code> object</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="gridPlotAssemblyVersions.html">gridPlotAssemblyVersions()</a></code> </p>
        </td>
        <td><p>Plot retrotransposon age distributions for predictions of different genome assembly versions</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="join.with.genome.tbl.html">join.with.genome.tbl()</a></code> </p>
        </td>
        <td><p>Join <code>gm_files</code> returned by <code>generate.multi.quality.filter.meta</code> with a genome information table</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="ltr.cn.html">ltr.cn()</a></code> </p>
        </td>
        <td><p>Copy Number Quantification of predicted LTRs (solo LTR prediction)</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="meta.apply.html">meta.apply()</a></code> </p>
        </td>
        <td><p>Apply functions to meta data generated by <code>LTRpred</code></p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="meta.seq.space.html">meta.seq.space()</a></code> </p>
        </td>
        <td><p>Quantification of the repeat space for multiple <code>Repeat Masker</code> files</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="meta.summarize.html">meta.summarize()</a></code> </p>
        </td>
        <td><p>Summarize (concatenate) all predictions of a <code>LTRpred.meta</code> run</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="motif.count.html">motif.count()</a></code> </p>
        </td>
        <td><p>Low level function to detect motifs in strings</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="plotCN.html">plotCN()</a></code> </p>
        </td>
        <td><p>Plot correlation between the copy number of the predicted 3' LTR and any other variable</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="plotCluster.html">plotCluster()</a></code> </p>
        </td>
        <td><p>Plot correlation between #cluster members and any other variable</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="plotDomain.html">plotDomain()</a></code> </p>
        </td>
        <td><p>Visualize the Protein Domain distribution of predicted LTR retrotransposons</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="plotFamily.html">plotFamily()</a></code> </p>
        </td>
        <td><p>Visualize the Superfamily distribution of predicted LTR retrotransposons</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="plotInterSpeciesCluster.html">plotInterSpeciesCluster()</a></code> </p>
        </td>
        <td><p>Plot inter species similarity between TEs (for a specific cluster)</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="plotLTRRange.html">plotLTRRange()</a></code> </p>
        </td>
        <td><p>Plot the location of putative LTR transposons along the chromosomes</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="plotLTRSim.html">plotLTRSim()</a></code> </p>
        </td>
        <td><p>Plot the age distribution of predicted LTR transposons</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="plotLTRSpace.html">plotLTRSpace()</a></code> </p>
        </td>
        <td><p>Genome Size vs. Protein Space</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="plotMainInterSpeciesCluster.html">plotMainInterSpeciesCluster()</a></code> </p>
        </td>
        <td><p>Plot inter species similarity between TEs (for the top 6 clusters)</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="plotProteinSpace.html">plotProteinSpace()</a></code> </p>
        </td>
        <td><p>Genome Size vs. Protein Space</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="plotRanges.html">plotRanges()</a></code> </p>
        </td>
        <td><p>Plot Ranges of an genomic feature</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="plotRepeatSpace.html">plotRepeatSpace()</a></code> </p>
        </td>
        <td><p>Genome Size vs. Repeat Space</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="plotRepeatSpaceCval.html">plotRepeatSpaceCval()</a></code> </p>
        </td>
        <td><p>Cvalue Genome Size vs. Repeat Space</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="plotSimCount.html">plotSimCount()</a></code> </p>
        </td>
        <td><p>Plot LTR Similarity vs. predicted LTR count</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="plotSize.html">plotSize()</a></code> </p>
        </td>
        <td><p>Genome size vs. LTR transposon content</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="plotSizeJumpers.html">plotSizeJumpers()</a></code> </p>
        </td>
        <td><p>Plot Genome size vs. LTR transposon count for jumpers</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="plotSizeMulti.html">plotSizeMulti()</a></code> </p>
        </td>
        <td><p>plotSize for <code>generate.multi.quality.filter.meta</code> output</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="plot_copynumber_individual.html">plot_copynumber_individual()</a></code> </p>
        </td>
        <td><p>Plot the median cluster copy number distribution of putative LTR transposons for and individual species</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="plot_copynumber_kingdom.html">plot_copynumber_kingdom()</a></code> </p>
        </td>
        <td><p>Plot the median cluster copy number distribution of putative LTR transposons for kingdoms of life</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="plot_copynumber_species.html">plot_copynumber_species()</a></code> </p>
        </td>
        <td><p>Plot the median cluster copy number distribution of putative LTR transposons for several species</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="plot_ltrwidth_individual.html">plot_ltrwidth_individual()</a></code> </p>
        </td>
        <td><p>Plot the width distribution of putative LTR transposons</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="plot_ltrwidth_kingdom.html">plot_ltrwidth_kingdom()</a></code> </p>
        </td>
        <td><p>Plot the median width distribution of putative LTR transposons for kingdoms of life</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="plot_ltrwidth_species.html">plot_ltrwidth_species()</a></code> </p>
        </td>
        <td><p>Plot the median width distribution of putative LTR transposons for several species</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="plot_pbswidth_kingdom.html">plot_pbswidth_kingdom()</a></code> </p>
        </td>
        <td><p>Plot the median PBS widths for kingdoms of life</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="pred2bed.html">pred2bed()</a></code> </p>
        </td>
        <td><p>Format LTR prediction data to BED file format</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="pred2csv.html">pred2csv()</a></code> </p>
        </td>
        <td><p>Format LTR prediction data to CSV file format</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="pred2fasta.html">pred2fasta()</a></code> </p>
        </td>
        <td><p>Save the sequence of the predicted LTR Transposons in a fasta file</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="pred2gff.html">pred2gff()</a></code> </p>
        </td>
        <td><p>Format LTR prediction data to GFF3 file format</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="pred2tsv.html">pred2tsv()</a></code> </p>
        </td>
        <td><p>Format LTR prediction data to TSV file format</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="quality.filter.html">quality.filter()</a></code> </p>
        </td>
        <td><p>Pipeline to eliminate false positive predictions of retrotransposons</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="quality.filter.meta.html">quality.filter.meta()</a></code> </p>
        </td>
        <td><p>Pipeline to eliminate false positive predictions of retrotransposons at a metagenomic scale</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="quant.cds.space.html">quant.cds.space()</a></code> </p>
        </td>
        <td><p>Quantify the coding sequence space within a genome</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="quant.protein.space.html">quant.protein.space()</a></code> </p>
        </td>
        <td><p>Quantify the protein space within a genome</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="quant.repeat.space.html">quant.repeat.space()</a></code> </p>
        </td>
        <td><p>Quantify the repeat space within a genome</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="quant.space.meta.html">quant.space.meta()</a></code> </p>
        </td>
        <td><p>Quantify the genomic loci space for multiple genomes, proteomes or Repeat Masker annotation files</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="read.blast6out.html">read.blast6out()</a></code> </p>
        </td>
        <td><p>Read file in blast6out format generated by USEARCH or VSEARCH</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="read.dfam.html">read.dfam()</a></code> </p>
        </td>
        <td><p>Import Dfam Query Output</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="read.ltrpred.html">read.ltrpred()</a></code> </p>
        </td>
        <td><p>Import <code>LTRpred</code> DataSheet</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="read.orfs.html">read.orfs()</a></code> </p>
        </td>
        <td><p>Read output of <code>ORFpred</code></p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="read.prediction.html">read.prediction()</a></code> </p>
        </td>
        <td><p>Import the output of LTRharvest or LTRdigest</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="read.seqs.html">read.seqs()</a></code> </p>
        </td>
        <td><p>Import sequences of predicted LTR transposons</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="read.tabout.html">read.tabout()</a></code> </p>
        </td>
        <td><p>Import information sheet returned by LTRdigest</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="read.uc.html">read.uc()</a></code> </p>
        </td>
        <td><p>Read file in USEARCH cluster format</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="rename.fasta.html">rename.fasta()</a></code> </p>
        </td>
        <td><p>Helper function to add species names to headers within fasta files</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="rename.fasta.meta.html">rename.fasta.meta()</a></code> </p>
        </td>
        <td><p>Meta function for applying <code>rename.fasta</code></p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="rename.organisms.html">rename.organisms()</a></code> </p>
        </td>
        <td><p>Rename file path names from <code>read_proteome</code>, etc.</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="repbase.clean.html">repbase.clean()</a></code> </p>
        </td>
        <td><p>Clean the initial Repbase database for BLAST</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="repbase.filter.html">repbase.filter()</a></code> </p>
        </td>
        <td><p>Filter the Repbase query output</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="repbase.query.html">repbase.query()</a></code> </p>
        </td>
        <td><p>Query the RepBase to annotate putative LTRs</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="tidy.datasheet.html">tidy.datasheet()</a></code> </p>
        </td>
        <td><p>Select most important columns of <code>LTRpred</code> output for further analytics</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="totalRepeatSpace.html">totalRepeatSpace()</a></code> </p>
        </td>
        <td><p>Quantify the total repeat space from Repeat Masker output in Mbp</p></td>
      </tr><tr>
        <!--  -->
        <td>
          <p><code><a href="ws.wrap.path.html">ws.wrap.path()</a></code> </p>
        </td>
        <td><p>Wrap whitespace in paths</p></td>
      </tr>
    </tbody>
    </table>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#section-all-functions">All functions</a></li>
    </ul>
  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by Hajk-Georg Drost.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
   </div>

  

  </body>
</html>

