<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Copy Number Quantification of predicted LTRs (solo LTR prediction) — ltr.cn • LTRpred</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Copy Number Quantification of predicted LTRs (solo LTR prediction) — ltr.cn" />
<meta property="og:description" content="Detect solo LTR copies and genomic locations of predicted LTR transposons using a BLAST search strategy." />
<meta name="twitter:card" content="summary" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">LTRpred</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/Introduction.html">Introduction to LTRpred</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/HajkD/LTRpred">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Copy Number Quantification of predicted LTRs (solo LTR prediction)</h1>
    <small class="dont-index">Source: <a href='https://github.com/HajkD/LTRpred/blob/master/R/ltr.cn.R'><code>R/ltr.cn.R</code></a></small>
    <div class="hidden name"><code>ltr.cn.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Detect solo LTR copies and genomic locations of predicted LTR transposons using a BLAST search strategy.</p>
    </div>

    <pre class="usage"><span class='fu'>ltr.cn</span>(
  <span class='no'>data.sheet</span>,
  <span class='no'>LTR.fasta_3ltr</span>,
  <span class='no'>LTR.fasta_5ltr</span>,
  <span class='no'>genome</span>,
  <span class='kw'>ltr.similarity</span> <span class='kw'>=</span> <span class='fl'>70</span>,
  <span class='kw'>scope.cutoff</span> <span class='kw'>=</span> <span class='fl'>0.85</span>,
  <span class='kw'>perc.ident.cutoff</span> <span class='kw'>=</span> <span class='fl'>70</span>,
  <span class='kw'>output</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>max.hits</span> <span class='kw'>=</span> <span class='fl'>500</span>,
  <span class='kw'>eval</span> <span class='kw'>=</span> <span class='fl'>1e-10</span>,
  <span class='kw'>cores</span> <span class='kw'>=</span> <span class='fl'>1</span>
)</pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>data.sheet</th>
      <td><p>path to the <code>*_LTRpred_DataSheet.csv</code> file generated by <code><a href='LTRpred.html'>LTRpred</a></code>.</p></td>
    </tr>
    <tr>
      <th>LTR.fasta_3ltr</th>
      <td><p>path to fasta file storing sequences of 3 prime LTRs generated by <code><a href='LTRpred.html'>LTRpred</a></code>.</p></td>
    </tr>
    <tr>
      <th>LTR.fasta_5ltr</th>
      <td><p>path to fasta file storing sequences of 5 prime LTRs generated by <code><a href='LTRpred.html'>LTRpred</a></code>.</p></td>
    </tr>
    <tr>
      <th>genome</th>
      <td><p>file path to the reference genome in which solo LTRs shall be found (in <code>fasta</code> format).</p></td>
    </tr>
    <tr>
      <th>ltr.similarity</th>
      <td><p>similarity threshold for defining LTR similarity.</p></td>
    </tr>
    <tr>
      <th>scope.cutoff</th>
      <td><p>similarity threshold for the scope variable. The scope of a BLAST hit is defined by <code>abs(s_len - alig_length) / s_len</code>
and aims to quantify the scope ('length similarity between hit and query') of the alignment. Default is <code>scope.cutoff = 0.85</code> meaning that
at least 85% of the length of the query sequence must match with the length of the subject sequence.</p></td>
    </tr>
    <tr>
      <th>perc.ident.cutoff</th>
      <td><p>choose the minimum sequence identity between the query sequence and subject hit sequence that have a scope &gt;= <code>scope.cutoff</code> (e.g. 0.85). 
Default is <code>perc.ident.cutoff = 70</code>. This threshold aims to detect BLAST hits that have similar sequence lengths 
(controlled by the <code>scope.cutoff</code> variable; e.g. <code>scope.cutoff = 0.85</code>) and within this scope at least e.g. 70% sequence identity (controlled by the <code>perc.ident.cutoff</code> variable).</p></td>
    </tr>
    <tr>
      <th>output</th>
      <td><p>file name of the BLAST output. If <code>output = NULL</code> (default) then the BLAST output file will be deleted after the result <code>data.frame</code> is returned by this function.</p></td>
    </tr>
    <tr>
      <th>max.hits</th>
      <td><p>maximum number of hits that shall be retrieved that still fulfill the e-value criterium.
Default is <code>max.hits = 65000</code>.</p></td>
    </tr>
    <tr>
      <th>eval</th>
      <td><p>e-value threshold for BLAST hit detection. Default is <code>eval = 1E-10</code>.</p></td>
    </tr>
    <tr>
      <th>cores</th>
      <td><p>number of cores for parallel computations.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="references"><a class="anchor" href="#references"></a>References</h2>

    <p>Altschul, S.F., Gish, W., Miller, W., Myers, E.W. &amp; Lipman, D.J. (1990) "Basic local alignment search tool." J. Mol. Biol. 215:403-410.</p>
<p>Gish, W. &amp; States, D.J. (1993) "Identification of protein coding regions by database similarity search." Nature Genet. 3:266-272.</p>
<p>Madden, T.L., Tatusov, R.L. &amp; Zhang, J. (1996) "Applications of network BLAST server" Meth. Enzymol. 266:131-141.</p>
<p>Altschul, S.F., Madden, T.L., Schaeffer, A.A., Zhang, J., Zhang, Z., Miller, W. &amp; Lipman, D.J. (1997) "Gapped BLAST and PSI-BLAST: a new generation of protein database search programs." Nucleic Acids Res. 25:3389-3402.</p>
<p>Zhang Z., Schwartz S., Wagner L., &amp; Miller W. (2000), "A greedy algorithm for aligning DNA sequences" J Comput Biol 2000; 7(1-2):203-14.</p>

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      <li><a href="#references">References</a></li>
    </ul>

    <h2>Author</h2>
    <p>Hajk-Georg Drost</p>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Hajk-Georg Drost.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


