<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html>
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Pipeline to eliminate false positive predictions of retrotransposons at a metagenomic scale — quality.filter.meta • LTRpred</title>

<!-- jquery -->
<script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/1.7.1/clipboard.min.js" integrity="sha384-cV+rhyOuRHc9Ub/91rihWcGmMmCXDeksTtCihMupQHSsi8GIIRDG0ThDc3HGQFJ3" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script>
<script src="../pkgdown.js"></script>



<meta property="og:title" content="Pipeline to eliminate false positive predictions of retrotransposons at a metagenomic scale — quality.filter.meta" />

<meta property="og:description" content="This function takes the file paths to the genomes folder and LTRpred.meta output folder as input and eliminates false positive retrotransposon predictions on a metagenomic scale." />
<meta name="twitter:card" content="summary" />


<!-- mathjax -->
<script src='https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">LTRpred</a>
        <span class="label label-default" data-toggle="tooltip" data-placement="bottom" title="Released package">1.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/Introduction.html">Introduction to LTRpred</a>
    </li>
  </ul>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/HajkD/LTRpred">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Pipeline to eliminate false positive predictions of retrotransposons at a metagenomic scale</h1>
    <small class="dont-index">Source: <a href='https://github.com/HajkD/LTRpred/blob/master/R/quality.filter.meta.R'><code>R/quality.filter.meta.R</code></a></small>
    <div class="hidden name"><code>quality.filter.meta.Rd</code></div>
    </div>

    <div class="ref-description">
    
    <p>This function takes the file paths to the genomes folder and <code><a href='LTRpred.meta.html'>LTRpred.meta</a></code> output folder as input and eliminates false positive retrotransposon predictions on a metagenomic scale.</p>
    
    </div>

    <pre class="usage"><span class='fu'>quality.filter.meta</span>(<span class='no'>kingdom</span>, <span class='no'>genome.folder</span>, <span class='no'>ltrpred.meta.folder</span>, <span class='no'>sim</span>,
  <span class='kw'>cut.range</span> <span class='kw'>=</span> <span class='fl'>2</span>, <span class='no'>n.orfs</span>, <span class='no'>strategy</span>, <span class='kw'>update</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>kingdom</th>
      <td><p>a character string specifying the kingdom of life to which genomes annotated with <code><a href='LTRpred.meta.html'>LTRpred.meta</a></code> belong to. E.g. <code>kingdom = "Plants"</code>. If annotates of a variety of kingdoms have been done users can for example specify <code>kingdom = "Various"</code>.</p></td>
    </tr>
    <tr>
      <th>genome.folder</th>
      <td><p>path to folder storing the genome assembly files that were used for <code><a href='LTRpred.meta.html'>LTRpred.meta</a></code> predictions.</p></td>
    </tr>
    <tr>
      <th>ltrpred.meta.folder</th>
      <td><p>path to folder storing the <code><a href='LTRpred.meta.html'>LTRpred.meta</a></code> output files.</p></td>
    </tr>
    <tr>
      <th>sim</th>
      <td><p>LTR similarity threshold. Only putative LTR transposons that fulfill this 
LTR similarity threshold will be retained.</p></td>
    </tr>
    <tr>
      <th>cut.range</th>
      <td><p>a numeric number indicating the interval size for binning LTR similarities.</p></td>
    </tr>
    <tr>
      <th>n.orfs</th>
      <td><p>minimum number of ORFs detected in the putative LTR transposon.</p></td>
    </tr>
    <tr>
      <th>strategy</th>
      <td><p>quality filter strategy. Options are</p><ul>
<li><p><code>strategy = "default"</code> : see section <code>Quality Control</code></p></li>
<li><p><code>strategy = "stringent"</code> : in addition to filter criteria specified in section <code>Quality Control</code>,
the filter criteria <code>!is.na(protein_domain)) | (dfam_target_name != "unknown")</code> is applied</p></li>
</ul></td>
    </tr>
    <tr>
      <th>update</th>
      <td><p>shall already existing <code>_SimilarityMatrix.csv</code> and <code>_GenomeInfo.csv</code> files be updated (<code>update = TRUE</code>) or can the already existing files be used (<code>update = FALSE</code>)?</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>A list with to list elements <code>sim_file</code> and <code>gm_file</code>. Each list element stores a <code>data.frame</code>:</p><ul>
<li><p><code>sim_file</code> (similarity file)</p><ul>
<li><p></p></li>
</ul></li>
<li><p><code>gm_file</code> (genome metrics file)</p><ul>
<li><p></p></li>
</ul></li>
</ul>

    
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p><strong>Quality Control</strong></p>
<ul>
<li><p><code>ltr.similarity</code>: Minimum similarity between LTRs. All TEs not matching this
 criteria are discarded.</p></li>
<li><p><code>n.orfs</code>: minimum number of Open Reading Frames that must be found between the
  LTRs. All TEs not matching this criteria are discarded.</p></li>
<li><p><code>PBS or Protein Match</code>: elements must either have a predicted Primer Binding
 Site or a protein match of at least one protein (Gag, Pol, Rve, ...) between their LTRs. All TEs not matching this criteria are discarded.</p></li>
<li><p>The relative number of N's (= nucleotide not known) in TE &lt;= 0.1. The relative number of N's is computed as follows: absolute number of N's in TE / width of TE.</p></li>
</ul>
    
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><p><code><a href='LTRpred.html'>LTRpred</a></code>, <code><a href='LTRpred.meta.html'>LTRpred.meta</a></code>, <code><a href='read.ltrpred.html'>read.ltrpred</a></code></p></div>
    

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>

      <li><a href="#details">Details</a></li>

      <li><a href="#see-also">See also</a></li>
          </ul>

    <h2>Author</h2>
    
Hajk-Georg Drost

  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by Hajk-Georg Drost.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
   </div>

  

  </body>
</html>

